% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\documentclass[
]{book}
\usepackage{xcolor}
\usepackage{amsmath,amssymb}
\setcounter{secnumdepth}{5}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\newsavebox\pandoc@box
\newcommand*\pandocbounded[1]{% scales image to fit in text height/width
  \sbox\pandoc@box{#1}%
  \Gscale@div\@tempa{\textheight}{\dimexpr\ht\pandoc@box+\dp\pandoc@box\relax}%
  \Gscale@div\@tempb{\linewidth}{\wd\pandoc@box}%
  \ifdim\@tempb\p@<\@tempa\p@\let\@tempa\@tempb\fi% select the smaller of both
  \ifdim\@tempa\p@<\p@\scalebox{\@tempa}{\usebox\pandoc@box}%
  \else\usebox{\pandoc@box}%
  \fi%
}
% Set default figure placement to htbp
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\usepackage[]{natbib}
\bibliographystyle{plainnat}
\usepackage{booktabs}

\usepackage{color}
\usepackage{framed}
\setlength{\fboxsep}{.8em}

% These colours were manually entered, they shouldn't matter unless you want pdf output

\newenvironment{redbox}{
  \definecolor{shadecolor}{RGB}{243, 154, 157}
  \color{white}
  \begin{shaded}}
 {\end{shaded}}

\newenvironment{bluebox}{
  \definecolor{shadecolor}{RGB}{172, 210, 237}
  \color{white}
  \begin{shaded}}
 {\end{shaded}}

\newenvironment{greenbox}{
  \definecolor{shadecolor}{RGB}{141, 181, 128}
  \color{white}
  \begin{shaded}}
 {\end{shaded}}
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Pathogen Genomic Epidemiology 2025},
  pdfauthor={Faculty: Angela McLaughlin, Emma Griffiths, Finlay Maguire, Gary Van Domselaar, Idowu Olawoye, Jennifer Guthrie, Charlie Barclay,},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Pathogen Genomic Epidemiology 2025}
\author{Faculty: Angela McLaughlin, Emma Griffiths, Finlay Maguire, Gary Van Domselaar, Idowu Olawoye, Jennifer Guthrie, Charlie Barclay,}
\date{November 24-26, 2025}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\part{Introduction}\label{part-introduction}

\chapter{Workshop Info}\label{workshop-info}

Welcome to the 2025 Pathogen Genomic Epidemiology Canadian Bioinformatics Workshop webpage!

\section{Pre-work}\label{pre-work}

\section{Class Photo}\label{class-photo}

Coming soon!

\section{Schedule}\label{schedule}

\chapter{Meet Your Faculty}\label{meet-your-faculty}

\subsubsection{Angela McLaughlin}\label{angela-mclaughlin}

\begin{quote}
Postdoctoral Fellow
Dalhousie University and University of Guelph
Burnaby, BC, Canada

--- \href{mailto:ez928230@dal.ca}{\nolinkurl{ez928230@dal.ca}}
\end{quote}

Angela McLaughlin is a postdoctoral fellow in Dr.~Finlay Maguire's lab at Dalhousie University, in collaboration with Dr.~Zvonimir Poljak at University of Guelph. Her research interests span viral phylogenetics (HIV-1, SARS-CoV-2, and influenza virus), genomic epidemiology, bioinformatics, public health, wildlife surveillance, and statistical/machine learning/mathematical models of pathogen transmission. Her current project aims to predict host specificity of avian influenza virus H5Nx using machine learning models of viral genomic features with phylogenetics-informed cross-validation and hierarchical segment to whole genome ensemble models.

\subsubsection{Emma Griffiths}\label{emma-griffiths}

\begin{quote}
Research Associate, Faculty of Health Sciences
Simon Fraser University
Vancouver, BC, Canada

--- \href{mailto:emma_griffiths@sfu.ca}{\nolinkurl{emma\_griffiths@sfu.ca}}
\end{quote}

Emma Griffiths is a research associate at the Centre for Infectious Disease Genomics and One Health (CIDGOH) in the Faculty of Health Sciences at Simon Fraser University in Vancouver, Canada. Her work focuses on developing and implementing ontologies and data standards for public health and food safety genomics to help improve data harmonization and integration. She is a member of the Standards Council of Canada and leads the Public Health Alliance for Genomic Epidemiology (PHA4GE) Data Structures Working Group.

\subsubsection{Finlay Maguire}\label{finlay-maguire}

\begin{quote}
Assistant Professor, Faculty of Computer Science and Department of Community Health \& Epidemiology,
Dalhousie University
Halifax, MS, Canada

--- \href{mailto:finlay.maguire@dal.ca}{\nolinkurl{finlay.maguire@dal.ca}}, finlaymagui.re
\end{quote}

Finlay Maguire is a genomic epidemiologist whose work centers on leveraging data in innovative ways to answer questions related to applied health and social issues. This includes developing bioinformatics methods to more effectively use genomic data to mitigate infectious diseases and broad interdisciplinary collaborations in areas such as refugee healthcare provision and online radicalisation. They are an active contributor to the national and international public health responses to emerging viral zoonoses and antimicrobial resistance, co-chair of the PHA4GE data structures working group, and act as a Pathogenomics Bioinformatics Lead for Sunnybrook's Shared Hospital Laboratory.

\subsubsection{Gary Van Domselaar}\label{gary-van-domselaar}

\begin{quote}
Chief, Bioinformatics, National Microbiology Laboratory
Public Health Agency of Canada
Winnipeg, MB, Canada

--- \href{mailto:gary.vandomselaar@phac-aspc.gc.ca}{\nolinkurl{gary.vandomselaar@phac-aspc.gc.ca}}
\end{quote}

Dr.~Gary Van Domselaar, PhD (University of Alberta, 2003) is the Chief of the Bioinformatics Section at the National Microbiology Laboratory in Winnipeg Canada and Associate Professor in the Department of Medical Microbiology and Infectious Diseases at the University of Manitoba. Dr.~Van Domselaar's lab develops bioinformatics methods and pipelines to understand, track, and control circulating infectious diseases in Canada and globally. His research and development activities span metagenomics, infectious disease genomic epidemiology, genome annotation, population structure analysis, and microbial genome wide association studies.

\subsubsection{Idowu Olawoye}\label{idowu-olawoye}

\begin{quote}
Postdoctoral Associate
University of Western Ontario
London, ON, Canada

--- \href{mailto:iolawoye@uwo.ca}{\nolinkurl{iolawoye@uwo.ca}}
\end{quote}

Idowu is a postdoctoral associate in the Guthrie Lab at the University of Western Ontario. His research focuses on utilizing computational biology to understand transmission patterns, genomic evolution, and antimicrobial resistance of bacterial pathogens. He has been involved in numerous bioinformatics workshops and trainings across Africa and most recently in Canada.

\subsubsection{Jennifer Guthrie}\label{jennifer-guthrie}

\begin{quote}
JOB TITLE
INSTITUTION
LOCATION

--- CONTACT INFO, IF PROVIDED
\end{quote}

BIO GOES HERE

\subsubsection{Charlie Barclay}\label{charlie-barclay}

\begin{quote}
MSc Graduate Student Researcher
University of British Columbia
Vancouver, BC, Canada

--- \href{mailto:cbarcl01@mail.ubc.ca}{\nolinkurl{cbarcl01@mail.ubc.ca}}
\end{quote}

Charlie is an ontology curator at the Centre for Infectious Disease Genomics and One Health (CIDGOH), focusing on data standards for contextual data of genomic epidemiology, including wastewater surveillance. With five years of experience in data management, specializing in biodiversity and genomics data, she also actively contributes to the Public Health Alliance for Genomic Epidemiology (PHA4GE) and the Global Alliance for Genomics and Health (GA4GH).

\chapter{Data and Compute Setup}\label{data-and-compute-setup}

\subsubsection{Course data downloads}\label{course-data-downloads}

Coming soon!

\subsubsection{Compute setup}\label{compute-setup}

Coming soon!

\chapter{Module 1 Data Curation and Data Sharing}\label{module-1-data-curation-and-data-sharing}

\section{Lecture}\label{lecture}

\section{Lab}\label{lab}

\chapter{Module 2 Emerging Pathogen Detection and Identification}\label{module-2-emerging-pathogen-detection-and-identification}

\section{Lecture}\label{lecture-1}

\section{Lab}\label{lab-1}

\chapter{Module 3 Pathogen Typing}\label{module-3-pathogen-typing}

\section{Lecture}\label{lecture-2}

\section{Lab}\label{lab-2}

\chapter{Module 4 Outbreak Analysis}\label{module-4-outbreak-analysis}

\section{Lecture}\label{lecture-3}

\section{Lab}\label{lab-3}

\chapter{Module 5 Phylodynamics}\label{module-5-phylodynamics}

\section{Lecture}\label{lecture-4}

\section{Lab}\label{lab-4}

\bibliography{book.bib,packages.bib}

\end{document}
