<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 2 Emerging Pathogen Detection and Identification | Pathogen Genomic Epidemiology 2025</title>
  <meta name="description" content="Workshop website for the 2025 Pathogen Genomic Epidemiology Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.41 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 2 Emerging Pathogen Detection and Identification | Pathogen Genomic Epidemiology 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/PGE_2025/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the 2025 Pathogen Genomic Epidemiology Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/PGE_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 2 Emerging Pathogen Detection and Identification | Pathogen Genomic Epidemiology 2025" />
  
  <meta name="twitter:description" content="Workshop website for the 2025 Pathogen Genomic Epidemiology Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/PGE_2025/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Angela McLaughlin, Emma Griffiths, Finlay Maguire, Gary Van Domselaar, Idowu Olawoye, Jennifer Guthrie, Zhibin Lu, Charlie Barclay," />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-1-data-curation-and-data-sharing.html"/>
<link rel="next" href="module-3-pathogen-typing.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "name": "Pathogen Genomic Epidemiology",
  "description": "The complete set of tutorials, slides, and practical exercises for the Pathogen Genomic Epidemiology 2025 Canadian Bioinformatics Workshop, hosted on GitHub Pages.",
  "url": "https://bioinformaticsdotca.github.io/PGE_2025",
  "learningResourceType": ["tutorial", "slides", "exercise"],
  "inLanguage": "en",
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "provider": {
    "@type": "Organization",
    "name": "Bioinformatics.ca",
    "url": "https://bioinformatics.ca"
  },
  "keywords": "",
  "isPartOf": {
    "@type": "Course",
    "@id": "https://bioinformatics.ca/workshops-all/2025-pathogen-genomic-epidemiology-online/"
  }
}
</script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br><br><br></a></strong></center></li>
<a href="./">
 <img src="img/pge-icon.png" width="70%" alt=" logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="chapter" data-level="" data-path="data-and-compute-setup.html"><a href="data-and-compute-setup.html"><i class="fa fa-check"></i>Data and Compute Setup</a></li>
<li class="chapter" data-level="" data-path="module-1-data-curation-and-data-sharing.html"><a href="module-1-data-curation-and-data-sharing.html"><i class="fa fa-check"></i>Module 1 Data Curation and Data Sharing</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-data-curation-and-data-sharing.html"><a href="module-1-data-curation-and-data-sharing.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1-data-curation-and-data-sharing.html"><a href="module-1-data-curation-and-data-sharing.html#lab"><i class="fa fa-check"></i>Lab</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html"><i class="fa fa-check"></i>Module 2 Emerging Pathogen Detection and Identification</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#copy-data-files"><i class="fa fa-check"></i>3.1. Copy data files</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#activate-environment"><i class="fa fa-check"></i>3.2. Activate environment</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#verify-your-workshop-machine-url"><i class="fa fa-check"></i>3.3. Verify your workshop machine URL</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html"><i class="fa fa-check"></i>Module 3 Pathogen Typing</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#lab-2"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#part-one-chewbbaca-objective"><i class="fa fa-check"></i>PART ONE: chewBBACA Objective</a></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#part-two-cgmlst-clustering-objective"><i class="fa fa-check"></i>PART TWO: cgMLST clustering Objective</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html"><i class="fa fa-check"></i>Module 4 Outbreak Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#lab-3"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#background"><i class="fa fa-check"></i>Background</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#set-up"><i class="fa fa-check"></i>Set-Up</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#cluster-identification"><i class="fa fa-check"></i>Cluster Identification</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#snp-analysis"><i class="fa fa-check"></i>SNP Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html"><i class="fa fa-check"></i>Module 5 Phylodynamics</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#lab-4"><i class="fa fa-check"></i>Lab</a></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="55" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="55" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=35% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<br><br>
<a href="https://uhn.ca/"><img src="img/sponsors/uhnlogo.png" width=55% style="display:inline-block;" style="padding-bottom: 20px" alt="University Health Network logo.">

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Pathogen Genomic Epidemiology 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-2-emerging-pathogen-detection-and-identification" class="section level1 hasAnchor">
<h1>Module 2 Emerging Pathogen Detection and Identification<a href="module-2-emerging-pathogen-detection-and-identification.html#module-2-emerging-pathogen-detection-and-identification" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-1" class="section level2 hasAnchor">
<h2>Lecture<a href="module-2-emerging-pathogen-detection-and-identification.html#lecture-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="lab-1" class="section level2 hasAnchor">
<h2>Lab<a href="module-2-emerging-pathogen-detection-and-identification.html#lab-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ol style="list-style-type: decimal">
<li><a href="#intro">Introduction</a></li>
<li><a href="#software">Software</a><br />
</li>
<li><a href="#setup">Setup</a></li>
<li><a href="#exercise">Exercise</a>
<ol style="list-style-type: decimal">
<li><a href="#exercise-background">Patient Background</a></li>
<li><a href="#exercise-overview">Overview</a></li>
<li><a href="#assembly-free">Assembly-free approach</a>
<ol style="list-style-type: decimal">
<li><a href="#exercise-examine-reads">Step 1: Examine the reads</a></li>
<li><a href="#exercise-quality-reads">Step 2: Clean and examine quality of the reads</a></li>
<li><a href="#exercise-host-filtering">Step 3: Host read filtering</a></li>
<li><a href="#exercise-classify-kraken">Step 4: Classify reads using Kraken2 database</a></li>
<li><a href="#exercise-pavian">Step 5: Generate an interactive html-based report using Pavian</a></li>
</ol></li>
<li><a href="#assembly-based">Assembly-based approach</a>
<ol style="list-style-type: decimal">
<li><a href="#exercise-metatranscriptomic-assembly">Step 6: Metatranscriptomic assembly</a></li>
<li><a href="#exercise-evaluate-assembly">Step 7: Evaluate assembly with Quast</a></li>
<li><a href="#exercise-blast">Step 8: Using BLAST to look for existing organisms</a></li>
</ol></li>
</ol></li>
<li><a href="#final">Final words</a></li>
</ol>
<p><a name="intro"></a>
### 1. Introduction</p>
<p>This tutorial aims to introduce a variety of software and concepts related to detecting emerging pathogens from a complex host sample. The provided data and methods are derived from real-world data, but have been modified to either illustrate a specific learning objective or to reduce the complexity of the problem. Contamination and a lack of large and accurate databases render detection of microbial pathogens difficult. As a disclaimer, all results produced from the tools described in this tutorial and others must also be verified with supplementary bioinformatics or wet-laboratory techniques.</p>
<p><a name="software"></a>
### 2. List of software for tutorial and its respective documentation</p>
<ul>
<li><a href="https://github.com/OpenGene/fastp">fastp</a></li>
<li><a href="https://multiqc.info/">multiqc</a></li>
<li><a href="https://kat.readthedocs.io/en/latest/">KAT</a></li>
<li><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a></li>
<li><a href="https://fbreitwieser.shinyapps.io/pavian/">Pavian</a></li>
<li><a href="https://github.com/voutcn/megahit">MEGAHIT</a></li>
<li><a href="http://quast.sourceforge.net/quast">Quast</a></li>
<li><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">NCBI blast</a></li>
</ul>
<p>The workshop machines already have this software installed within a conda environment, but to perform this analysis later on, you can make use of the conda environment located at <a href="environment.yml">environment.yml</a> to install the necessary software.</p>
<p><a name="setup"></a>
### 3. Exercise setup</p>
<div id="copy-data-files" class="section level3 hasAnchor">
<h3>3.1. Copy data files<a href="module-2-emerging-pathogen-detection-and-identification.html#copy-data-files" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To begin, we will copy over the exercises to <code>~/workspace</code>. This let’s use view the resulting output files in a web browser.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb1-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> ~/CourseData/module2/module2_workspace/ ~/workspace/</span>
<span id="cb1-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb1-2" tabindex="-1"></a><span class="bu">cd</span> ~/workspace/module2_workspace/analysis</span></code></pre></div>
<p>When you are finished with these steps you should be inside the directory <code>/home/ubuntu/workspace/module2_workspace/analysis</code>. You can verify this by running the command <code>pwd</code>.</p>
<p><strong>Output after running <code>pwd</code></strong></p>
<pre><code>/home/ubuntu/workspace/module2_workspace/analysis</code></pre>
<p>You should also have a directory like <code>data/</code> one directory up from here. To check this, you can run <code>ls ../</code>:</p>
<p><strong>Output after running <code>ls ../</code></strong></p>
<pre><code>analysis  data  precomputed-analysis</code></pre>
</div>
<div id="activate-environment" class="section level3 hasAnchor">
<h3>3.2. Activate environment<a href="module-2-emerging-pathogen-detection-and-identification.html#activate-environment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Next we will activate the <a href="https://docs.conda.io/en/latest/">conda</a> environment, which will have all the tools needed by this tutorial pre-installed. To do this please run the following:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb4-1" tabindex="-1"></a><span class="ex">conda</span> activate module2-emerging-pathogen</span></code></pre></div>
<p>You should see the command-prompt (where you type commands) switch to include <code>(module2-emerging-pathogen)</code> at the beginning, showing you are inside this environment. You should also be able to run one of the commands like <code>kraken2 --version</code> and see output:</p>
<p><strong>Output after running <code>kraken2 --version</code></strong></p>
<pre><code>Kraken version 2.17.1
Copyright 2013-2023, Derrick Wood (dwood@cs.jhu.edu)</code></pre>
</div>
<div id="verify-your-workshop-machine-url" class="section level3 hasAnchor">
<h3>3.3. Verify your workshop machine URL<a href="module-2-emerging-pathogen-detection-and-identification.html#verify-your-workshop-machine-url" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This exercise will produce output files intended to be viewed in a web browser. These should be accessible by going to <a href="http://xx.uhn-hpc.ca" class="uri">http://xx.uhn-hpc.ca</a> in your web browser where <strong>xx</strong> is your particular number (like 01, 02, etc). If you are able to view a list of files and directories, try clicking the link for <strong>module2_workspace</strong>. This page will be referred to later to view some of our output files. In addition, the link <strong>precompuated-analysis</strong> will contain all the files we will generate during this lab.</p>
<p><a name="exercise"></a>
### 4. Exercise</p>
<p><a name="exercise-background"></a>
### 4.1. Patient Background:</p>
<p>A 41-year-old man was admitted to a hospital 6 days after the onset of disease. He reported fever, chest tightness, unproductive cough, pain and weakness. Preliminary investigations excluded the presence of influenza virus, <em>Chlamydia pneumoniae</em>, <em>Mycoplasma pneumoniae</em>, and other common respiratory pathogens. After 3 days of treatment the patient was admitted to the intensive care unit, and 6 days following admission the patient was transferred to another hospital.</p>
<p>To further investigate the cause of illness, a sample of bronchoalveolar lavage fluid (BALF) was collected from the patient and metatranscriptomic sequencing was performed (that is, the RNA from the sample was sequenced). In this lab, you will examine the metatranscriptomic data using a number of bioinformatics methods and tools to attempt to identify the cause of the illness.</p>
<p><em>Note: The patient information and data was derived from a real study (shown at the end of the lab).</em></p>
<p><a name="exercise-overview"></a>
### 4.2. Overview</p>
<p>We will proceed through the following steps to attempt to diagnose the situation.</p>
<ul>
<li>Trim and clean sequence reads using <code>fastp</code></li>
<li>Filter host (human) reads with <code>kat</code></li>
<li>Run Kraken2 with a bacterial and viral database to look at the taxonomic makeup of the reads.</li>
<li>Assemble the metatranscriptome with <code>megahit</code></li>
<li>Examine assembly quality using <code>quast</code> and possible pathogens with <code>blast</code></li>
</ul>
<hr />
<p><a name="assembly-free"></a>
### Assembly-free approach</p>
<p>The first set of steps follows through an assembly-free approach where we will perform taxonomic classification of the reads without constructing a metagenomics assembly.</p>
<p><a name="exercise-examine-reads"></a>
#### Step 1: Examine the reads</p>
<p>Let’s first take a moment to examine the reads from the metatranscrimptomic sequencing. Note that for metatranscriptomic sequencing, while we are sequencing the RNA, this was performed by first generating complementary DNA (cDNA) to the RNA and sequencing the cDNA. Hence you will see thymine (T) instead of uracil (U) in the sequence data.</p>
<p>The reads were generated from paired-end sequencing, which means that a particular fragment (of cDNA) was sequenced twice–once from either end (see the <a href="https://www.illumina.com/science/technology/next-generation-sequencing/plan-experiments/paired-end-vs-single-read.html">Illumina Paired vs. Single-End Reads</a> for some additional details). These pairs of cDNA sequence reads are stored as separate files (named <code>emerging-pathogen-reads_1.fastq.gz</code> and <code>emerging-pathogen-reads_2.fastq.gz</code>). You can see each file by running <code>ls</code>:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb6-1" tabindex="-1"></a><span class="fu">ls</span> ../data</span></code></pre></div>
<p><strong>Output</strong></p>
<pre><code>emerging-pathogen-reads_1.fastq.gz  emerging-pathogen-reads_2.fastq.gz</code></pre>
<p>We can look at the contents of one of the files by running <code>less</code> (you can look at the other pair of reads too, but it will look very similar):</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb8-1" tabindex="-1"></a><span class="fu">less</span> ../data/emerging-pathogen-reads_1.fastq.gz</span></code></pre></div>
<p><strong>Output</strong></p>
<pre><code>@SRR10971381.5 5 length=151
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@SRR10971381.7 7 length=151
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@SRR10971381.33 33 length=115
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@SRR10971381.56 56 length=151
CCCGTGTTCGATTGGCATTTCACCCCTATCCACAACTCATCCCAAAGCTTTTCAACGCTCACGAGTTCGGTCCTCCACACAATTTTACCTGTGCTTCAACCTGGCCATGGATAGATCACTACGGTTTCGGGTCTACTATTACTAACTGAAC
+
FFFFFFFAFFFFFFAFFFFFF6FFFFFFFFF/FFFFFFFFFFFF/FFFFFFFFFFFFFFFFFAFFFFFFFFFFFAFFFFF/FFAF/FAFFFFFFFFFAFFFF/FFFFFFFFFFF/F=FF/FFFFA6FAFFFFF//FFAFFFFFFAFFFFFF</code></pre>
<p>These reads are in the <a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a>, which stores a single read as a block of 4 lines: <strong>identifier</strong>, <strong>sequence</strong>, <strong>+ (separator)</strong>, <strong>quality scores</strong>. In this file, we can see a lot of lines with <code>NNN...</code> for the sequence letters, which means that these portions of the read are not determined. We will remove some of these undetermined (and uninformative) reads in the next step.</p>
<p><a name="exercise-quality-reads"></a>
#### Step 2: Clean and examine quality of the reads</p>
<p>As we saw from looking at the data, reads that come directly off of a sequencer may be of variable quality which might impact the downstream analysis. We will use the software <a href="https://github.com/OpenGene/fastp">fastp</a> to both clean and trim reads (removing poor-quality reads or sequencing adapters) as well as examine the quality of the reads. To do this please run the following (the expected time of this command is shown as <code># Time: 30 seconds</code>).</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb10-1" tabindex="-1"></a><span class="co"># Time: 30 seconds</span></span>
<span id="cb10-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb10-2" tabindex="-1"></a><span class="ex">fastp</span> <span class="at">--detect_adapter_for_pe</span> <span class="at">--in1</span> ../data/emerging-pathogen-reads_1.fastq.gz <span class="at">--in2</span> ../data/emerging-pathogen-reads_2.fastq.gz <span class="at">--out1</span> cleaned_1.fastq <span class="at">--out2</span> cleaned_2.fastq</span></code></pre></div>
<p>You should see the following as output:</p>
<p><strong>Output</strong></p>
<pre><code>Detecting adapter sequence for read1...
No adapter detected for read1

Detecting adapter sequence for read2...
No adapter detected for read2
[...]
Insert size peak (evaluated by paired-end reads): 150

JSON report: fastp.json
HTML report: fastp.html

fastp --detect_adapter_for_pe --in1 ../data/emerging-pathogen-reads_1.fastq.gz --in2 ../data/emerging-pathogen-reads_2.fastq.gz --out1 cleaned_1.fastq --out2 cleaned_2.fastq
fastp v0.24.0, time used: 17 seconds</code></pre>
<div id="examine-output" class="section level4 hasAnchor">
<h4>Examine output<a href="module-2-emerging-pathogen-detection-and-identification.html#examine-output" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You should now be able to nagivate to &lt; <a href="http://xx.uhn-hpc.ca/module2_workspace/analysis" class="uri">http://xx.uhn-hpc.ca/module2_workspace/analysis</a>&gt; and see some of the output files. In particular, you should be able to find <strong>fastp.html</strong>, which contains a report of the quality of the reads and how many were removed. Please take a look at this report now:</p>
<p><img src="img/module2/fastp.png" alt="fastp" width="750" /></p>
<p>This should show an overview of the quality of the reads before and after filtering with <code>fastp</code>. Using this report, please answer the following questions.</p>
</div>
<div id="step-2-questions" class="section level4 hasAnchor">
<h4>Step 2: Questions<a href="module-2-emerging-pathogen-detection-and-identification.html#step-2-questions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li>Looking at the <strong>Filtering result</strong> section, how many reads <strong>passed filters</strong>? How many were removed due to <strong>low quality</strong>? How many were removed due to <strong>too many N</strong>?</li>
<li>Looking at the <strong>Adapters</strong> section, were there many adapters that needed to be trimmed in this data?</li>
<li>Compare the <strong>quality</strong> and <strong>base contents</strong> plots <strong>Before filtering</strong> and <strong>After filtering</strong>? How do they differ?</li>
</ol>
<hr />
<p><a name="exercise-host-filtering"></a>
#### Step 3: Host read filtering</p>
<p>The next step is to remove any host reads (in this case Human reads) from our dataset as we are not focused on examining host reads. There are several different tools that can be used to filter out host reads such as Bowtie2 or KAT. In this demonstration, we have selected to run KAT followed by Kraken2, but you could likely accomplish something similar by using Bowtie2 followed by Kraken2.</p>
<p>Command documentation is available <a href="http://kat.readthedocs.io/en/latest/using.html#sequence-filtering">here</a></p>
<p>KAT works by breaking down each read into small fragements of length <em>k</em>, k-mers, and compares them to a k-mer database of the human reference genome. Subsequently, the complete read is either assigned into a matched or unmatched (filtered) file if 10% of the k-mers in the read have been found in the human database.</p>
<p><img src="img/module2/kat.png" alt="kat" width="750" /></p>
<p>Let’s run KAT now.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb12-1" tabindex="-1"></a><span class="co"># Time: 3 minutes</span></span>
<span id="cb12-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb12-2" tabindex="-1"></a><span class="ex">kat</span> filter seq <span class="at">-t</span> 4 <span class="at">-i</span> <span class="at">-o</span> filtered <span class="at">--seq</span> cleaned_1.fastq <span class="at">--seq2</span> cleaned_2.fastq ~/CourseData/module2/db/kat_db/human_kmers.jf</span></code></pre></div>
<p>The arguments for this command are:</p>
<ul>
<li><code>filter seq</code>: Specifies that we are running a specific subcommand to <strong>filter sequences</strong>.</li>
<li><code>-t 4</code>: The number of threads to use (we have 4 CPU cores on these machines so we are using 4 threads).</li>
<li><code>--seq --seq2</code> arguments to provide corresponding forward and reverse fastq reads (the cleaned reads from <code>fastp</code>)</li>
<li><code>-i</code>: Inverts the filter, that is we wish to output sequences <strong>not found</strong> in the human kmer database to a file.</li>
<li><code>-o filtered</code> Provide prefix for all files generated by the command. In our case, we will have two output files <strong>filtered.in.R1.fastq</strong> and <strong>filetered.in.R2.fastq</strong>.</li>
<li><code>~/CourseData/module2/db/kat_db/human_kmers.jf</code> the human k-mer database we’re using made with jellyfish.</li>
</ul>
<p>As the command is running you should see the following output on your screen:</p>
<p><strong>Output</strong></p>
<pre><code>Kmer Analysis Toolkit (KAT) V2.4.2

Running KAT in filter sequence mode
-----------------------------------

Loading hashes into memory... done.  Time taken: 40.7s

Filtering sequences ...
Processed 100000 pairs
Processed 200000 pairs
[...]
Finished filtering.  Time taken: 130.5s

Found 1127908 / 1306231 to keep

KAT filter seq completed.
Total runtime: 182.8s</code></pre>
<p>If the command was successful, your current directory should contain two new files:</p>
<ul>
<li><code>filtered.in.R1.fastq</code></li>
<li><code>filtered.in.R2.fastq</code></li>
</ul>
<p>These are the set of reads minus any reads that matched the human genome. The message <code>Found 1127908 / 1306231 to keep</code> tells us how many read-pairs were kept (the number in the <code>filtered.in.*.fastq</code> files) vs. the total number of read-pairs.</p>
<hr />
<p><a name="exercise-classify-kraken"></a>
#### Step 4: Classify reads using Kraken2 database</p>
<p>Now that we have most, if not all, host reads filtered out, it’s time to classify the remaining reads to identify the likely taxonomic category they belong to.</p>
<p>Database selection is one of the most crucial parts of running Kraken. One of the many factors that must be considered is the computational resources available. Our current AWS image for the course has only 16G of memory. A major disadvantage of Kraken2 is that it loads the entire database into memory. With the <a href="https://benlangmead.github.io/aws-indexes/k2">standard viral, bacterial, and archael database</a> on the order of 50 GB we would be unable to run the full database on the course machine. To help mitigate this, Kraken2 allows reduced databases to be constructed, which will still give reasonable results but we may end up with more reads unclassified or classified at a higher taxonomic rank. We have constructed our own smaller Kraken2 database using only bacterial, human, and viral data. We will be using this database.</p>
<p>Lets run the following command in our current directory to classify our reads against the Kraken2 database.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb14-1" tabindex="-1"></a><span class="co"># Time: 1 minute</span></span>
<span id="cb14-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb14-2" tabindex="-1"></a><span class="ex">kraken2</span> <span class="at">--db</span> ~/CourseData/module2/db/kraken2_db <span class="at">--threads</span> 4 <span class="at">--paired</span> <span class="at">--output</span> kraken_out.txt <span class="at">--report</span> kraken_report.txt <span class="at">--unclassified-out</span> kraken2_unclassified#.fastq filtered.in.R1.fastq filtered.in.R2.fastq</span></code></pre></div>
<p>This should produce output similar to below:</p>
<p><strong>Output</strong></p>
<pre><code>Loading database information... done.
1127908 sequences (315.54 Mbp) processed in 5.332s (12693.3 Kseq/m, 3551.00 Mbp/m).
  880599 sequences classified (78.07%)
  247309 sequences unclassified (21.93%)</code></pre>
</div>
<div id="examine-kraken_report.txt" class="section level4 hasAnchor">
<h4>Examine <code>kraken_report.txt</code><a href="module-2-emerging-pathogen-detection-and-identification.html#examine-kraken_report.txt" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s examine the text-based report of Kraken2:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb16-1" tabindex="-1"></a><span class="fu">less</span> kraken_report.txt</span></code></pre></div>
<p>This should produce output similar to the following:</p>
<pre><code> 21.93  247309  247309  U       0       unclassified
 78.07  880599  30      R       1       root
 78.01  879899  124     R1      131567    cellular organisms
 76.37  861411  19285   D       2           Bacteria
 56.53  637572  2       D1      1783270       FCB group
 56.53  637558  1571    D2      68336           Bacteroidetes/Chlorobi group
 56.39  635982  1901    P       976               Bacteroidetes
 55.10  621496  35      C       200643              Bacteroidia
 55.09  621417  19584   O       171549                Bacteroidales
 53.18  599872  2464    F       171552                  Prevotellaceae
 52.96  597396  397538  G       838                       Prevotella
  4.74  53473   53473   S       28137                       Prevotella veroralis
  4.34  48940   48940   S       1177574                     Prevotella jejuni
  2.56  28880   470     G1      2638335                     unclassified Prevotella</code></pre>
<p>This will show the top taxonomic ranks (right-most column) as well as the percent and number of reads that fall into these categories (left-most columns). For example:</p>
<ul>
<li>The very first row <code>21.93  247309  247309  U       0       unclassified</code> shows us that <strong>247309 (21.93%)</strong> of the reads processed by Kraken2 are unclassified (remember we only used a database containing bacterial, viral, and human representatives).</li>
<li>The 4th line <code>76.37  861411  19285   D       2           Bacteria</code> tells us that <strong>861411 (76.37%)</strong> of our reads fall into the <strong>Bacteria</strong> domain (the <code>D</code> in the fourth column is the taxonomic rank, <strong><code>D</code>omain</strong>). The number <code>19285</code> tells us that <code>19285</code> of the reads are assigned directly to the <strong>Bacteria</strong> domain but cannot be assigned to any lower taxonomic rank (they match with too many diverse types of bacteria).</li>
</ul>
<p>More details about how to read this report can be found at <a href="https://github.com/DerrickWood/kraken2/wiki/Manual#sample-report-output-format" class="uri">https://github.com/DerrickWood/kraken2/wiki/Manual#sample-report-output-format</a>. In the next step we will represent this data visually as a multi-layered pie chart.</p>
</div>
<div id="examine-kraken_out.txt" class="section level4 hasAnchor">
<h4>Examine <code>kraken_out.txt</code><a href="module-2-emerging-pathogen-detection-and-identification.html#examine-kraken_out.txt" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s also take a look at <code>kraken_out.txt</code>. This file contains the kraken2 results, but divided up into a classification for every read.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb18-1" tabindex="-1"></a><span class="ex">column</span> <span class="at">-s</span><span class="st">$&#39;</span><span class="dt">\t</span><span class="st">&#39;</span> <span class="at">-t</span> kraken_out.txt <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span></code></pre></div>
<p><code>column</code> formats a text file (<code>kraken_out.txt</code>) into multiple columns according to a tab delimiter character (flag <code>-s'$\t'</code>) and produces a table (flag <code>-t</code>).</p>
<p><strong>Output</strong></p>
<pre><code>C       SRR10971381.56  29465   151|151 0:40 909932:2 0:8 909932:2 0:20 29465:4 0:7 178327&gt;
C       SRR10971381.97  838     122|122 0:44 2:5 0:23 838:1 0:10 838:2 0:3 |:| 0:3 838:2 0&gt;
C       SRR10971381.126 9606    109|109 0:2 9606:5 0:7 9606:1 0:12 9606:1 0:47 |:| 0:47 96&gt;
C       SRR10971381.135 838     151|151 0:95 838:3 0:19 |:| 0:15 838:1 0:12 838:5 0:6 838:&gt;
C       SRR10971381.219 1177574 151|151 0:11 838:3 0:5 838:2 0:9 838:5 0:11 838:1 976:5 83&gt;
C       SRR10971381.223 838     151|151 0:117 |:| 0:61 838:4 0:40 838:1 0:11
[...]</code></pre>
<p>This shows us a taxonomic classification for every read (one read per line). For example:</p>
<ul>
<li>On the first line, <code>C  SRR10971381.56        29465</code> tells us that this read with identifier <code>SRR10971381.56</code> is classified <code>C</code> (matches to something in the Kraken2 database) and matches to the taxonomic category <code>29465</code>, which is the NCBI taxonomy identifer. In this case <code>29465</code> corresponds to the <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=29465&amp;lvl=3&amp;lin=f&amp;keep=1&amp;srchmode=1&amp;unlock"><em>Veillonella</em></a>.</li>
</ul>
<p>More information on interpreting this file can be found at <a href="https://github.com/DerrickWood/kraken2/wiki/Manual#standard-kraken-output-format" class="uri">https://github.com/DerrickWood/kraken2/wiki/Manual#standard-kraken-output-format</a>.</p>
<hr />
<p><a name="exercise-pavian"></a>
#### Step 5: Generate an interactive html-based report using Pavian</p>
<p>Instead of reading text-based files like above, we can visualize this information using <a href="https://fbreitwieser.shinyapps.io/pavian/">Pavian</a>, which can be used to construct an interactive summary and visualization of metagenomics data. Pavian supports a number of metagenomics analysis software outputs, including Kraken/Kraken2. To visualize the Kraken2 output we just generated, we can upload the <code>kraken_report.txt</code> file to the web application. Please do this now using the following steps:</p>
<ol style="list-style-type: decimal">
<li><p>Download the <code>kraken_report.txt</code> to your local machine from <a href="http://xx.uhn-hpc.ca/module2_workspace/analysis" class="uri">http://xx.uhn-hpc.ca/module2_workspace/analysis</a> (you can right-click and select <strong>Save as…</strong> on the file).</p></li>
<li><p>Visit the <a href="https://fbreitwieser.shinyapps.io/pavian/">Pavian</a> website and click on <strong>Upload files &gt; Browse…</strong> and select the file <code>kraken_report.txt</code> we just downloaded.</p>
<p><img src="img/module2/pavian-upload.png" alt="pavian-upload" width="750" /></p></li>
<li><p>Select <strong>Generate HTML report …</strong> to generate the Pavian report.</p>
<p><img src="img/module2/pavian-generate-report.png" alt="pavian-generate" width="300" /></p></li>
<li><p>Open the generated report HTML file in your web browser.</p></li>
</ol>
<p>If all the steps are completed successfully then the report you should see should look like the following:</p>
<p><img src="img/module2/pavian-report.png" alt="pavian-report" width="750" /></p>
</div>
<div id="step-5-questions" class="section level4 hasAnchor">
<h4>Step 5: Questions<a href="module-2-emerging-pathogen-detection-and-identification.html#step-5-questions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li>What are the percentages of <strong>Unclassified</strong>, <strong>Microbial</strong>, <strong>Bacterial</strong>, <strong>Viral</strong>, <strong>Fungal</strong>, and <strong>Protozoan</strong> reads in this dataset?</li>
<li>Scroll down to the <strong>Classification results</strong> section of the report and flip through the <strong>Bacteria</strong>, <strong>Viruses</strong>, and <strong>Eukaryotes</strong> tabs. What is the top organism in each of these three categories and how many reads?</li>
<li>This data was derived from RNA (instead of DNA) and some viruses are RNA-based. If we focus in on the <strong>Viruses</strong> category, is there anything here that could be consistent with the patient’s symptoms?</li>
<li>Given the results of Pavian, can you form a hypothesis as to the cause of the patient’s symptoms?</li>
</ol>
<hr />
<p><a name="assembly-based"></a>
### Assembly-based approach</p>
<p><a name="exercise-metatranscriptomic-assembly"></a>
#### Step 6: Metatranscriptomic assembly</p>
<p>In order to investigate the data further we will assemble the metatranscriptome using the software <a href="https://github.com/voutcn/megahit">MEGAHIT</a>. What this will do is integrate all the read data together to attempt to produce the longest set of contiguous sequences possible (contigs). To do this please run the following:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb20-1" tabindex="-1"></a><span class="co"># Time: 6 minutes</span></span>
<span id="cb20-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb20-2" tabindex="-1"></a><span class="ex">megahit</span> <span class="at">-t</span> 4 <span class="at">-1</span> filtered.in.R1.fastq <span class="at">-2</span> filtered.in.R2.fastq <span class="at">-o</span> megahit_out</span></code></pre></div>
<p>If everything is working you should expect to see the following as output:</p>
<p><strong>Output</strong></p>
<pre><code>2025-11-19 15:45:30 - MEGAHIT v1.2.9
2025-11-19 15:45:30 - Using megahit_core with POPCNT and BMI2 support
2025-11-19 15:45:30 - Convert reads to binary library
2025-11-19 15:45:31 - b&#39;INFO  sequence/io/sequence_lib.cpp  :   75 - Lib 0 (/media/cbwdata/workspace/module2_workspace/analysis/filtered.in.R1.fastq,/media/cbwdata/workspace/module2_workspace/analysis/filtered.in.R2.fastq): pe, 2255816 reads, 151 max length&#39;
2025-11-19 15:45:32 - b&#39;INFO  utils/utils.h                 :  152 - Real: 1.3294\tuser: 1.2124\tsys: 0.3011\tmaxrss: 166748&#39;
2025-11-19 15:45:32 - k-max reset to: 141
2025-11-19 15:45:32 - Start assembly. Number of CPU threads 4
2025-11-19 15:45:32 - k list: 21,29,39,59,79,99,119,141
2025-11-19 15:45:32 - Memory used: 14741121024
2025-11-19 15:45:32 - Extract solid (k+1)-mers for k = 21

[...]

2025-11-19 15:49:08 - Assemble contigs from SdBG for k = 141
2025-11-19 15:49:09 - Merging to output final contigs
2025-11-19 15:49:09 - 3112 contigs, total 1536607 bp, min 203 bp, max 29867 bp, avg 493 bp, N50 463 bp
2025-11-19 15:49:09 - ALL DONE. Time elapsed: 218.408143 seconds</code></pre>
<p>Once everything is completed, you will have a directory <code>megahit_out/</code> with the output. Let’s take a look at this now:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb22-1" tabindex="-1"></a><span class="fu">ls</span> megahit_out/</span></code></pre></div>
<p><strong>Output</strong></p>
<pre><code>checkpoints.txt  done  final.contigs.fa  intermediate_contigs  log  options.json</code></pre>
<p>It’s specifically the <strong>final.contigs.fa</strong> file that contains our metatranscriptome assembly. This will contain the largest <em>contiguous</em> sequences MEGAHIT was able to construct from the sequence reads. We can look at the contents with the command <code>head</code> (<code>head</code> prints the first 10 lines of a file):</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb24-1" tabindex="-1"></a><span class="fu">head</span> megahit_out/final.contigs.fa</span></code></pre></div>
<p><strong>Output</strong></p>
<pre><code>&gt;k141_0 flag=1 multi=3.0000 len=312
ATACTGATCTTAGAAAGCTTAGATTTCATCTTTTCAATTGGTGTATCGAATTTAGATACAAATTTAGCTAAGGATTTAGACATTTCAGCTTTATCTACAGTAGAGTATACTTTAATATCTTGAAGTACACCAGTTACTTTAGACTTAATCAAAATTTTACCCAAATCATTAACTAGATCTTTAGAATCAGAATTCTTTTCTACCATTTTAGCGATGATATCTGTTGCATCTTGATCTTCAAATGAAGATCTATATGACATGATAGTTTGACCTTCTTGTAGTTGAGATCCAACTTCTAAACATTCGATGTCT
&gt;k141_785 flag=1 multi=6.0000 len=355
TATAGAGATAGGTGATAAGGTTTTCCTAGGCGCACAATCAGGTGTACCTGGAAGTTTGAAGGCTAATCAACAACTGATTGGTACACCTCCAATGGAGCAACGCTCTTACTTTAAGTCGCAAGCCATCTTCCGCAGACTGCCTGAGATGTACAAGCAACTAAGCGACTTGCAGAAAGAAATAGACCAATTGAAAAAGAACAGATAACATTCATGGATACAATAAAACAGAAGACATTAAAAGGGAGTTTCTCGCTCTTTGGCAAGGGACTTCACACAGGTTTGAGTCTTACCGTGACGTTTAATCCGGCTCCAGAGAATACTGGCTATAAGATACAACGTATAGATCTTGAGGGGC
[...]</code></pre>
<p>It can be a bit difficult to get an overall idea of what is in this file, so in the next step we will use the software <a href="http://quast.sourceforge.net/quast">Quast</a> to summarize the assembly information.</p>
<hr />
<p><a name="exercise-evaluate-assembly"></a>
#### Step 7: Evaluate assembly with Quast</p>
<p><a href="http://quast.sourceforge.net/quast">Quast</a> can be used to provide summary statistics on the output of assembly software. Quast will take as input an assembled genome or metagenome (a FASTA file of different sequences) and will produce HTML and PDF reports. We will run Quast on our data by running the following command:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb26-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb26-1" tabindex="-1"></a><span class="co"># Time: 2 seconds</span></span>
<span id="cb26-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb26-2" tabindex="-1"></a><span class="ex">quast</span> <span class="at">-t</span> 4 megahit_out/final.contigs.fa</span></code></pre></div>
<p>You should expect to see the following as output:</p>
<p><strong>Output</strong></p>
<pre><code>/home/ubuntu/.conda/envs/module2-emerging-pathogen/bin/quast -t 4 megahit_out/final.contigs.fa

Version: 5.3.0

System information:
  OS: Linux-6.14.0-1016-aws-x86_64-with-glibc2.39 (linux_64)
  Python version: 3.9.23
  CPUs number: 4

Started: 2025-11-19 15:50:27

[...]

Finished: 2025-11-19 15:50:28
Elapsed time: 0:00:01.273336
NOTICEs: 1; WARNINGs: 0; non-fatal ERRORs: 0

Thank you for using QUAST!</code></pre>
<p>Quast writes it’s output to a directory <code>quast_results/</code>, which includes HTML and PDF reports. We can view this using a web browser by navigating to <a href="http://xx.uhn-hpc.ca/module2_workspace/analysis/" class="uri">http://xx.uhn-hpc.ca/module2_workspace/analysis/</a> and clicking on <strong>quast_results</strong> then <strong>latest</strong> then <strong>icarus.html</strong>. From here, click on <strong>Contig size viewer</strong>. You should see the following:</p>
<p><img src="img/module2/quast-contigs.png" alt="quast" width="750" /></p>
<p>This shows the length of each contig in the <code>megahit_out/final.contigs.fa</code> file, sorted by size.</p>
<p>(If it is having issues you can also see the needed information in the <code>pdf</code> report)</p>
</div>
<div id="step-7-questions" class="section level4 hasAnchor">
<h4>Step 7: Questions<a href="module-2-emerging-pathogen-detection-and-identification.html#step-7-questions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li>What is the length of the largest contig in the genome? How does it compare to the length of the 2nd and 3rd largest contigs?</li>
<li>Given that this is RNASeq data (i.e., sequences derived from RNA), what is the most common type of RNA you should expect to find? What are the approximate lengths of these RNA fragments? Is the largest contig an outlier (i.e., is it much longer than you would expect)?</li>
<li>Is there another type of source for this RNA fragment that could explain it’s length? Possibly a <a href="https://en.wikipedia.org/wiki/Coronavirus#Genome">Virus</a>?</li>
<li>Also try looking at the QUAST report (<a href="http://xx.uhn-hpc.ca/module2_workspace/analysis/quast_results/latest/" class="uri">http://xx.uhn-hpc.ca/module2_workspace/analysis/quast_results/latest/</a> then clicking on <strong>report.html</strong>). How many contigs &gt;= 1000 bp are there compared to the number &lt; 1000 bp?</li>
</ol>
<hr />
<p><a name="exercise-blast"></a>
#### Step 8: Use BLAST to look for existing organisms</p>
<p>In order to get a better handle on what the identity of the largest contigs could be, let’s use <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to compare to a database of existing viruses. Please run the following:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb28-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb28-1" tabindex="-1"></a><span class="co"># Time: 1 second</span></span>
<span id="cb28-2"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb28-2" tabindex="-1"></a><span class="ex">seqkit</span> sort <span class="at">--by-length</span> <span class="at">--reverse</span> megahit_out/final.contigs.fa <span class="kw">|</span> <span class="ex">seqkit</span> head <span class="at">-n</span> 50 <span class="op">&gt;</span> contigs-50.fa</span>
<span id="cb28-3"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb28-3" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-db</span> ~/CourseData/module2/db/blast_db/ref_viruses_rep_genomes_modified <span class="at">-query</span> contigs-50.fa <span class="at">-html</span> <span class="at">-out</span> blast_results.html</span></code></pre></div>
<p>As output you should see something like (<code>blastn</code> won’t print any output):</p>
<p><strong>Output</strong></p>
<pre><code>[INFO] read sequences ...
[INFO] 3112 sequences loaded
[INFO] sorting ...
[INFO] output ...</code></pre>
<p>Here, we first use <a href="https://bioinf.shenwei.me/seqkit/">seqkit</a> to sort all contigs by length with the largest ones first (<code>seqkit sort --by-length --reverse ...</code>) and we then extract only the top <strong>50</strong> longest contigs (<code>seqkit head -n 50</code>) and write these to a file <strong>contigs-50.fa</strong> (<code>&gt; contigs-50.fa</code>).</p>
<p><em>Note that the pipe <code>|</code> character will take the output of one command (<code>seqkit sort --by-length ...</code>, which sorts sequences in the file by length) and forward it into the input of another command (<code>seqkit head -n 50</code>, which takes only the first 50 sequences from the file). The greater-than symbol <code>&gt;</code> takes the output of one command <code>seqkit head ...</code> and writes it to a file (named <code>contigs-50.fa</code>).</em></p>
<p>The next command will run <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> on these top 50 longest contigs using a pre-computed database of viral genomes (<code>blastn -db ~/CourseData/module2/db/blast_db/ref_viruses_rep_genomes_modified -query contigs-50.fa ...</code>). The <code>-html -out blast_results.html</code> tells BLAST to write its results as an HTML file.</p>
<p>To view these results, please browse to <a href="http://xx.uhn-hpc.ca/module2_workspace/analysis/blast_results.html" class="uri">http://xx.uhn-hpc.ca/module2_workspace/analysis/blast_results.html</a> to view the ouptut <code>blast_results.html</code> file. This should look something like below:</p>
<p><img src="img/module2/blast-report.png" alt="blast" width="750" /></p>
</div>
<div id="step-8-questions" class="section level4 hasAnchor">
<h4>Step 8: Questions<a href="module-2-emerging-pathogen-detection-and-identification.html#step-8-questions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>What is the closest match for the longest contig you find in your data? What is the percent identify for this match (the value Z in <code>Identities = X/Y (Z%)</code>). Recall that if a pathogen is an emerging/novel pathogen then you may not get a perfect match to any existing organisms.</p></li>
<li><p>Using the BLAST report alongside all other information we’ve gathered, what can you say about what pathogen may be causing the patient’s symptoms?</p></li>
<li><p>It can be difficult to examine all the contigs/BLAST matches at once with the standard BLAST report (which shows the full alignment). We can modify the BLAST command to output a tab-separated file, with one BLAST HSP (a high-scoring segment pair) per line. To do this please run the following:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb30-1"><a href="module-2-emerging-pathogen-detection-and-identification.html#cb30-1" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-db</span> ~/CourseData/module2/db/blast_db/ref_viruses_rep_genomes_modified <span class="at">-query</span> contigs-50.fa <span class="at">-outfmt</span> <span class="st">&#39;7 qseqid length slen pident sseqid stitle&#39;</span> <span class="at">-out</span> blast_report.tsv</span></code></pre></div>
<p>This should construct a tabular BLAST report with the columns labeled like <code>query id, alignment length, subject length, % identity, subject id, subject title</code>. Taking a look at the file <code>blast_report.tsv</code>, what are all the different BLAST matches you can find (the different values for <code>subject title</code>)? How do they compare in terms of <code>% identity</code> and <code>alignment length</code> (in general, higher values for both of these should be better matches)?</p></li>
</ol>
<hr />
<p><a name="final"></a>
### 5. Final words</p>
<p>Congratulations, you’ve finished this lab. As a final check on your results, you can use <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">NCBI’s online tool</a> to perform a BLAST on our top 50 contigs to see what matches to the contigs.</p>
<p>The source of the data and patient background information can be found at <a href="https://doi.org/10.1038/s41586-020-2008-3" class="uri">https://doi.org/10.1038/s41586-020-2008-3</a> (<strong>clicking this link will reveal what the illness is</strong>). The only modification made to the original metatranscriptomic reads was to reduce them to 10% of the orginal file size.</p>
<p>While we used <strong>MEGAHIT</strong> to perform the assembly, there are a number of other more recent assemblers that may be useful. In particular, the <a href="https://cab.spbu.ru/software/spades/">SPAdes</a> suite of tools (such as <a href="https://doi.org/10.1093/bioinformatics/btaa490">metaviralspades</a> or <a href="https://doi.org/10.1093/gigascience/giz100">rnaspades</a>) may be useful to look into for this sort of data analysis.</p>
<p>As a final note, NCBI also performs taxonomic analysis using their own software and you can actually view these using Krona directly from NCBI. Please click <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR10971381">here</a> and go to the <em>Analysis</em> tab for NCBI’s taxonomic analysis of this sequence data (<strong>clicking this link will reveal what the illness is</strong>).</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-1-data-curation-and-data-sharing.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-3-pathogen-typing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/bioinformaticsdotca//blob/main/020-module-2.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
