<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 5 Phylodynamics | IDE Module 5 Lab</title>
  <meta name="description" content="Phylodynamics of SARS-CoV-2 B.1.1.7" />
  <meta name="generator" content="bookdown 0.45 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 5 Phylodynamics | IDE Module 5 Lab" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/PGE_2025/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Phylodynamics of SARS-CoV-2 B.1.1.7" />
  <meta name="github-repo" content="bioinformaticsdotca/PGE_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 5 Phylodynamics | IDE Module 5 Lab" />
  
  <meta name="twitter:description" content="Phylodynamics of SARS-CoV-2 B.1.1.7" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/PGE_2025/img/bioinformatics_logo.png" />

<meta name="author" content="Angela McLaughlin and Finlay Maguire" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-4-outbreak-analysis.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "name": "Pathogen Genomic Epidemiology",
  "description": "The complete set of tutorials, slides, and practical exercises for the Pathogen Genomic Epidemiology 2025 Canadian Bioinformatics Workshop, hosted on GitHub Pages.",
  "url": "https://bioinformaticsdotca.github.io/PGE_2025",
  "learningResourceType": ["tutorial", "slides", "exercise"],
  "inLanguage": "en",
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "provider": {
    "@type": "Organization",
    "name": "Bioinformatics.ca",
    "url": "https://bioinformatics.ca"
  },
  "keywords": "",
  "isPartOf": {
    "@type": "Course",
    "@id": "https://bioinformatics.ca/workshops-all/2025-pathogen-genomic-epidemiology-online/"
  }
}
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br><br><br></a></strong></center></li>
<a href="./">
 <img src="img/pge-icon.png" width="70%" alt=" logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="chapter" data-level="" data-path="data-and-compute-setup.html"><a href="data-and-compute-setup.html"><i class="fa fa-check"></i>Data and Compute Setup</a></li>
<li class="chapter" data-level="" data-path="module-1-data-curation-and-data-sharing.html"><a href="module-1-data-curation-and-data-sharing.html"><i class="fa fa-check"></i>Module 1 Data Curation and Data Sharing</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-data-curation-and-data-sharing.html"><a href="module-1-data-curation-and-data-sharing.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1-data-curation-and-data-sharing.html"><a href="module-1-data-curation-and-data-sharing.html#lab"><i class="fa fa-check"></i>Lab</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html"><i class="fa fa-check"></i>Module 2 Emerging Pathogen Detection and Identification</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#intro"><i class="fa fa-check"></i>1. Introduction</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#software"><i class="fa fa-check"></i>2. List of software for tutorial and its respective documentation</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#setup"><i class="fa fa-check"></i>3. Exercise setup</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#exercise"><i class="fa fa-check"></i>4. Exercise</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#assembly-free"><i class="fa fa-check"></i>Assembly-free approach</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#assembly-based"><i class="fa fa-check"></i>Assembly-based approach</a></li>
<li class="chapter" data-level="" data-path="module-2-emerging-pathogen-detection-and-identification.html"><a href="module-2-emerging-pathogen-detection-and-identification.html#final"><i class="fa fa-check"></i>5. Final words</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html"><i class="fa fa-check"></i>Module 3 Pathogen Typing</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#lab-2"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#part-one-chewbbaca-objective"><i class="fa fa-check"></i>PART ONE: chewBBACA Objective</a></li>
<li class="chapter" data-level="" data-path="module-3-pathogen-typing.html"><a href="module-3-pathogen-typing.html#part-two-cgmlst-clustering-objective"><i class="fa fa-check"></i>PART TWO: cgMLST clustering Objective</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html"><i class="fa fa-check"></i>Module 4 Outbreak Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#lab-3"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#background"><i class="fa fa-check"></i>Background</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#set-up"><i class="fa fa-check"></i>Set-Up</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#cluster-identification"><i class="fa fa-check"></i>Cluster Identification</a></li>
<li class="chapter" data-level="" data-path="module-4-outbreak-analysis.html"><a href="module-4-outbreak-analysis.html#snp-analysis"><i class="fa fa-check"></i>SNP Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html"><i class="fa fa-check"></i>Module 5 Phylodynamics</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#lab-4"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#table-of-contents"><i class="fa fa-check"></i>Table of contents</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#introduction-1"><i class="fa fa-check"></i>1. Introduction</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#list-of-software-for-tutorial"><i class="fa fa-check"></i>2. List of software for tutorial</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#exercise-setup"><i class="fa fa-check"></i>3. Exercise setup</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#maximum-likelihood-phylodynamic-analysis"><i class="fa fa-check"></i>4. Maximum likelihood phylodynamic analysis</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#visualizing-the-phylodynamic-analysis"><i class="fa fa-check"></i>5. Visualizing the phylodynamic analysis</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#delphy-analysis"><i class="fa fa-check"></i>6. Delphy analysis</a></li>
<li class="chapter" data-level="" data-path="module-5-phylodynamics.html"><a href="module-5-phylodynamics.html#end-of-lab"><i class="fa fa-check"></i>7. End of lab</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="55" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="55" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=35% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<br><br>
<a href="https://uhn.ca/"><img src="img/sponsors/uhnlogo.png" width=55% style="display:inline-block;" style="padding-bottom: 20px" alt="University Health Network logo.">

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">IDE Module 5 Lab</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-5-phylodynamics" class="section level1 hasAnchor">
<h1>Module 5 Phylodynamics<a href="module-5-phylodynamics.html#module-5-phylodynamics" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-4" class="section level2 hasAnchor">
<h2>Lecture<a href="module-5-phylodynamics.html#lecture-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="lab-4" class="section level2 hasAnchor">
<h2>Lab<a href="module-5-phylodynamics.html#lab-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="table-of-contents" class="section level3 hasAnchor">
<h3>Table of contents<a href="module-5-phylodynamics.html#table-of-contents" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li><a href="#intro-mod5">Introduction</a></li>
<li><a href="#software-mod5">Software</a><br />
</li>
<li><a href="#setup-mod5">Exercise setup</a></li>
<li><a href="#phylodyn-mod5">Maximum likelihood phylodynamics analysis</a></li>
<li><a href="#visualize-tree-mod5">Visualizing the phylogeny in auspice</a></li>
<li><a href="#delphy-mod5">Delphy Bayesian phylodynamic analysis</a></li>
<li><a href="#end-mod5">End of lab</a></li>
</ol>
<p><a name="intro-mod5"></a></p>
</div>
<div id="introduction-1" class="section level3 hasAnchor">
<h3>1. Introduction<a href="module-5-phylodynamics.html#introduction-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="background-1" class="section level4 hasAnchor">
<h4>1.1. Background<a href="module-5-phylodynamics.html#background-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>SARS-CoV-2 Pango lineage B.1.1.7 <a href="https://www.nature.com/articles/s41564-020-0770-5">(Rambaut et al. 2020)</a> was the first officially designated variant of concern (VOC) by the World Health Organization <a href="https://www.nature.com/articles/s41564-021-00932-w">(Konings et al. 2021)</a>, due to its increased transmissibility compared to other circulating lineages. Although B.1.1.7 was discovered in December 2020, it originated in the months preceding that. For a more thorough investigation into the circumstances of B.1.1.7 discovery and emergence, see <a href="https://academic.oup.com/ve/article/8/2/veac080/6677185">(Hill et al., 2022)</a> on “The origins and molecular evolution of SARS-CoV-2 lineage B.1.1.7 in the UK”.</p>
<p>Evaluating when and where (and in what host, for multi-host pathogens) novel lineages or variants with increased transmissibility or virulence emerge is important towards improving our understanding of the conditions supporting the emergence of variants against which pharmaceutical (vaccines, therapies) or non-pharmaceutical interventions may be less effective.</p>
<p>In this tutorial, we will analyze early SARS-CoV-2 B.1.1.7 sequences and associated metadata from the first months following its detection to reconstruct its likely time and place of origin using phylodynamics and phylogeography, in both maximum likelihood and Bayesian statistical frameworks.</p>
</div>
<div id="objectives-of-this-analysis" class="section level4 hasAnchor">
<h4>1.2. Objectives of this analysis:<a href="module-5-phylodynamics.html#objectives-of-this-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><p>Download and clean publicly available SARS-CoV-2 B.1.1.7 sequences</p></li>
<li><p>Remove sequences with incomplete sample dates and subsample fewer sequences (keeping earliest with higher probability)</p></li>
<li><p>Align sequences and filter out low-quality sequences</p></li>
<li><p>Summarize the spatiotemporal distribution of samples</p></li>
<li><p>Infer a maximum likelihood phylogenetic tree (divergence tree)</p></li>
<li><p>Infer a maximum likelihood time-scaled tree using a molecular clock model</p></li>
<li><p>Conduct ancestral state reconstruction to identify the most likely geographic origin of lineage B.1.1.7</p></li>
<li><p>Compare the maximum likelihood estimates to those obtained in a Bayesian framework using Delphy</p></li>
</ul>
</div>
<div id="what-statistical-approach-is-best" class="section level4 hasAnchor">
<h4>1.3. What statistical approach is best?<a href="module-5-phylodynamics.html#what-statistical-approach-is-best" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Maximum likelihood and Bayesian approaches to phylodynamics and phylogeography differ in their assumptions, algorithms, computational speed, scalability, and interpretation.</p>
<p>For large scale analyses (&gt;10,000 sequences), maximum likelihood pipelines implemented in augur, treetime, iqtree, and R package ape, are more feasible, while analyses of smaller datasets (up to ~1000 sequences) can be done using Bayesian frameworks implemented in <a href="https://delphy.fathom.info/">Delphy</a> and <a href="http://www.beast2.org/">BEAST</a>, with the added benefits of incorporating uncertainty in the tree inference in parameter estimation and of specifying priors to limit the search space.</p>
<p>If you are interested in exploring more complex applications and methods for phylodynamic analyses, I recommend checking out resources from the <a href="https://taming-the-beast.org/workshops/Taming-the-BEAST-Online/">“Taming the BEAST” workshops</a> that are held semi-regularly, as well as he tutorials shared by the <a href="http://www.beast2.org/tutorials/">BEAST Developers</a>. Paul Lewis’ <a href="http://phyloseminar.org/recorded.html">phyloseminar.org</a> lectures on Bayesian phylogenetics (<a href="https://www.youtube.com/watch?v=4PWlnNsfz90">part 1</a> and <a href="https://www.youtube.com/watch?v=TLtOS--YwkU">part 2</a>) are also an excellent place to start to learn the basic theory of Bayesian methods.</p>
<p><em>Let’s go.</em></p>
<p><a name="software-mod5"></a></p>
</div>
</div>
<div id="list-of-software-for-tutorial" class="section level3 hasAnchor">
<h3>2. List of software for tutorial<a href="module-5-phylodynamics.html#list-of-software-for-tutorial" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><a href="https://tidyverse.org/packages/">R package tidyverse</a></li>
<li><a href="https://github.com/Bioconductor/Biostrings">R package Biostrings</a></li>
<li><a href="https://cran.r-project.org/web/packages/gtools/index.html">R package gtools</a></li>
<li><a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html">R package RColorBrewer</a></li>
<li><a href="https://docs.nextstrain.org/projects/augur/en/stable/index.html">augur</a></li>
<li><a href="https://mafft.cbrc.jp/alignment/software/">mafft</a></li>
<li><a href="http://www.iqtree.org/">iqtree</a></li>
<li><a href="https://treetime.readthedocs.io/en/latest/">treetime</a></li>
<li><a href="https://auspice.us/">Auspice</a></li>
<li><a href="https://delphy.fathom.info/">Delphy</a></li>
</ul>
<p><a name="setup-mod5"></a></p>
</div>
<div id="exercise-setup" class="section level3 hasAnchor">
<h3>3. Exercise setup<a href="module-5-phylodynamics.html#exercise-setup" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="data-source" class="section level4 hasAnchor">
<h4>3.0. Data source<a href="module-5-phylodynamics.html#data-source" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We will not spend time downloading the data, but rather will copy the datafiles as described in the subsequent section. However, it’s important to understand how this was done for the sake of reproducibility and interpretation.</p>
</div>
<div id="filters-applied-to-download-this-dataset" class="section level4 hasAnchor">
<h4>Filters applied to download this dataset<a href="module-5-phylodynamics.html#filters-applied-to-download-this-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>SARS-CoV-2 B.1.1.7 (Alpha variant) data from <a href="https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/">NCBI Virus</a></p>
<ul>
<li>Pango lineage: B.1.1.7</li>
<li>Host: human</li>
<li>Dates:
<ul>
<li>Collection date: Sep 1 2020 to Jan 1 2021 (First few months of detection)</li>
<li>Release date: Sep 1 2020 to Jan 1 2022 (allow one year delayed submission)</li>
</ul></li>
<li>Sequence Quality:
<ul>
<li>Nucleotide completeness: complete only</li>
</ul></li>
<li>Nucleotide &gt; Download all results</li>
<li>Data type
<ul>
<li><ol style="list-style-type: decimal">
<li>sequence data = nucleotide; Fasta definition line: accession, country, collection date</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Results table in csv format; Results csv &gt; Accession with version number</li>
</ol></li>
</ul></li>
<li>N=15790 fasta sequences (b117_seq.fasta) and metadata (b117_seq.csv)</li>
</ul>
<p><strong>Information to include when downloading metadata from NCBI:</strong>
<img src="img/PGE_mod5/ncbi_down.png" alt="ncbi_download" /></p>
</div>
<div id="copy-data-files-1" class="section level4 hasAnchor">
<h4>3.1. Copy data files<a href="module-5-phylodynamics.html#copy-data-files-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To begin, we will copy over the downloaded files to <code>~/workspace</code>.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb67"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb67-1"><a href="module-5-phylodynamics.html#cb67-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> ~/CourseData/module5 ~/workspace/.</span>
<span id="cb67-2"><a href="module-5-phylodynamics.html#cb67-2" tabindex="-1"></a><span class="bu">cd</span> ~/workspace/module5</span></code></pre></div>
<p>When you are finished with these steps you should be inside the directory <code>/home/ubuntu/workspace/module5</code>. You can verify this by running the command <code>pwd</code>.</p>
<p><strong>Output after running <code>pwd</code></strong></p>
<pre><code>/home/ubuntu/workspace/module5</code></pre>
<p>You should also see a directory <code>data/</code> in the current directory which contains all the input data. You can verify this by running <code>ls data/download</code>:</p>
<p><strong>Output after running <code>ls data/download</code></strong></p>
<pre><code>b117_seq.csv    b117_seq.fasta     reference_MN908947.3.fasta </code></pre>
<p>It is also good to verify the data was copied fully. Run this code to count how many sequences are in the fasta file (each one starts with &gt;):
<strong>Commands</strong></p>
<div class="sourceCode" id="cb70"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb70-1"><a href="module-5-phylodynamics.html#cb70-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">&quot;^&gt;&quot;</span> data/download/b117_seq.fasta </span></code></pre></div>
<p>This should be 15790.</p>
</div>
<div id="activate-conda-environment" class="section level4 hasAnchor">
<h4>3.2. Activate conda environment<a href="module-5-phylodynamics.html#activate-conda-environment" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Next we will activate the <a href="https://docs.conda.io/en/latest/">conda</a> environment, which will have all the tools needed by this tutorial pre-installed.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb71"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb71-1"><a href="module-5-phylodynamics.html#cb71-1" tabindex="-1"></a><span class="ex">conda</span> activate sc2b117_phylodyn</span></code></pre></div>
<p>You should see the command-prompt (where you type commands) switch to include <code>(sc2b117_phylodyn)</code> at the beginning, showing you are inside this environment. You should also be able to run the <code>augur</code> command like <code>augur --version</code> and see output:</p>
<p><strong>Output after running <code>augur --version</code></strong></p>
<pre><code>augur 32.0.0 </code></pre>
</div>
<div id="find-your-ip-address-to-setup-an-aws-instance" class="section level4 hasAnchor">
<h4>3.3. Find your IP address to setup an AWS instance<a href="module-5-phylodynamics.html#find-your-ip-address-to-setup-an-aws-instance" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Similar to yesterday, we will want to either use the assigned hostname (e.g., xx.uhn-hpc.ca where xx is your instance number) or find the IP address of your machine on AWS so we can access some files from your machine on the web browser. To find your IP address you can run:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb73"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb73-1"><a href="module-5-phylodynamics.html#cb73-1" tabindex="-1"></a><span class="ex">curl</span> http://checkip.amazonaws.com</span></code></pre></div>
<p>This should print a number like XX.XX.XX.XX. Once you have your address, try going to <a href="http://xx.uhn-hpc.ca" class="uri">http://xx.uhn-hpc.ca</a> and clicking the link for <strong>module5</strong>. This page will be referred to later to view some of our output files. In addition, the link <strong>precomputed-analysis</strong> will contain all the files we will generate during this lab (phylogenetic trees, etc).</p>
</div>
<div id="inspect-and-clean-data" class="section level4 hasAnchor">
<h4>3.4. Inspect and clean data<a href="module-5-phylodynamics.html#inspect-and-clean-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The first step of any phylogenetic analysis is to inspect and clean the dataset. Initial cleaning (before even looking at the sequence quality) steps include checking that accessions and sequence names are complete and unique, checking the completeness of sampling dates and locations (especially in phylodynamic analyses where these play a central role), checking the range of dates are realistic, and looking at how many geographic categories there are and whether some are more well-represented than others.</p>
<p>Run the following Rscript in the command line to de-duplicate seqeunces, remove incomplete dates, and subsample the data.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb74"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb74-1"><a href="module-5-phylodynamics.html#cb74-1" tabindex="-1"></a><span class="ex">Rscript</span> scripts/dedup_comp_dates.R <span class="st">&quot;data/download/b117_seq.csv&quot;</span> <span class="st">&quot;data/download/b117_seq.fasta&quot;</span></span></code></pre></div>
<p>While this is running, let’s have a look inside the R script. Open the .Rproj file in RStudio, then open scripts/dedup_comp_dates.R. Note that while we have called this script from the command line, you can also run it within Rstudio by manually specifying the arguments.</p>
<p>What is happening inside this script:</p>
<ul>
<li>De-duplicate identical sequences (taking earliest collection date of replicates)</li>
<li>Rename sequences without any spaces and add sequence name to metadata</li>
<li>Identify and remove incomplete collection date (yyyy or yyyy-mm)</li>
<li>Create a smaller subsampled dataset ‘complete_sampUK’ with equal number of non-UK and UK samples
<ul>
<li><em>NOTE: this is the dataset that we will be working with in this lab due to limited time.</em> In a more robust analysis, we would include more samples, repeat sampling with different seeds, and rather than sampling at random, we might consider sampling proportionally to geographies’ diagnosed cases through time to more accurately reflect spatiotemporal dynamics</li>
</ul></li>
<li>Export fasta and csv files into data/complete subfolder</li>
</ul>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>How many sequences were downloaded?</li>
<li>How many sequences were there after de-duplicating?</li>
<li>After de-duplicating, how many had complete dates? How many were sampled (sampUK)?</li>
<li>How might you modify the subsampling scheme to reduce bias? to quantify uncertainty in the sampling process?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>n=15790</li>
<li>n=14774</li>
<li>n=10813; n=178</li>
<li>Proportional to case counts through time and space. Multiple
subsamples of different sizes or proportional contributions of
countries.</li>
</ul></div></details>
<p>Once that code is done, let’s run some R scripts from the command line to quickly summarize the number and/or proportion of sequences by country through time. Alternatively, in RStudio, go to file&gt;new file&gt;Rscript, copy/paste code chunk, save and run it there.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb75"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb75-1"><a href="module-5-phylodynamics.html#cb75-1" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb75-2"><a href="module-5-phylodynamics.html#cb75-2" tabindex="-1"></a><span class="ex">R</span> </span>
<span id="cb75-3"><a href="module-5-phylodynamics.html#cb75-3" tabindex="-1"></a></span>
<span id="cb75-4"><a href="module-5-phylodynamics.html#cb75-4" tabindex="-1"></a><span class="ex">library</span><span class="er">(</span><span class="ex">tidyverse</span><span class="kw">)</span></span>
<span id="cb75-5"><a href="module-5-phylodynamics.html#cb75-5" tabindex="-1"></a><span class="ex">library</span><span class="er">(</span><span class="ex">RColorBrewer</span><span class="kw">)</span></span>
<span id="cb75-6"><a href="module-5-phylodynamics.html#cb75-6" tabindex="-1"></a></span>
<span id="cb75-7"><a href="module-5-phylodynamics.html#cb75-7" tabindex="-1"></a><span class="co">#data in</span></span>
<span id="cb75-8"><a href="module-5-phylodynamics.html#cb75-8" tabindex="-1"></a><span class="ex">dat.samp</span><span class="op">&lt;</span>-read.delim<span class="er">(</span><span class="st">&quot;data/complete/b117_seq_complete_sampUK.tsv&quot;</span><span class="kw">)</span></span>
<span id="cb75-9"><a href="module-5-phylodynamics.html#cb75-9" tabindex="-1"></a><span class="ex">nrow</span><span class="er">(</span><span class="ex">dat.samp</span><span class="kw">)</span></span>
<span id="cb75-10"><a href="module-5-phylodynamics.html#cb75-10" tabindex="-1"></a><span class="ex">dat.samp</span><span class="va">$date</span><span class="op">&lt;</span>-as.Date<span class="er">(</span><span class="ex">dat.samp</span><span class="va">$date</span><span class="kw">)</span></span>
<span id="cb75-11"><a href="module-5-phylodynamics.html#cb75-11" tabindex="-1"></a></span>
<span id="cb75-12"><a href="module-5-phylodynamics.html#cb75-12" tabindex="-1"></a><span class="co">#range of dates</span></span>
<span id="cb75-13"><a href="module-5-phylodynamics.html#cb75-13" tabindex="-1"></a><span class="ex">range</span><span class="er">(</span><span class="ex">dat.samp</span><span class="va">$date</span><span class="kw">)</span></span>
<span id="cb75-14"><a href="module-5-phylodynamics.html#cb75-14" tabindex="-1"></a></span>
<span id="cb75-15"><a href="module-5-phylodynamics.html#cb75-15" tabindex="-1"></a><span class="co">#countries </span></span>
<span id="cb75-16"><a href="module-5-phylodynamics.html#cb75-16" tabindex="-1"></a><span class="ex">t.c</span><span class="op">&lt;</span>-table<span class="er">(</span><span class="ex">dat.samp</span><span class="va">$Country</span><span class="kw">)</span> <span class="ex">%</span><span class="op">&gt;</span>% sort<span class="er">(</span><span class="ex">decreasing</span> = T<span class="kw">);</span><span class="ex">t.c</span></span>
<span id="cb75-17"><a href="module-5-phylodynamics.html#cb75-17" tabindex="-1"></a><span class="ex">order.c</span><span class="op">&lt;</span>-names<span class="er">(</span><span class="ex">t.c</span><span class="kw">)</span> <span class="co">#order of countries w/ most to least seqs/samples</span></span>
<span id="cb75-18"><a href="module-5-phylodynamics.html#cb75-18" tabindex="-1"></a><span class="ex">n.c</span><span class="op">&lt;</span>-length<span class="er">(</span><span class="ex">unique</span><span class="er">(</span><span class="ex">dat.samp</span><span class="va">$Country</span><span class="kw">));</span><span class="ex">n.c</span></span>
<span id="cb75-19"><a href="module-5-phylodynamics.html#cb75-19" tabindex="-1"></a></span>
<span id="cb75-20"><a href="module-5-phylodynamics.html#cb75-20" tabindex="-1"></a><span class="co">#more specific geos also available</span></span>
<span id="cb75-21"><a href="module-5-phylodynamics.html#cb75-21" tabindex="-1"></a><span class="ex">table</span><span class="er">(</span><span class="ex">dat.samp</span><span class="va">$Geo_Location</span><span class="kw">)</span> <span class="ex">%</span><span class="op">&gt;</span>% sort<span class="er">(</span><span class="ex">decreasing</span> = T<span class="kw">)</span></span>
<span id="cb75-22"><a href="module-5-phylodynamics.html#cb75-22" tabindex="-1"></a><span class="ex">length</span><span class="er">(</span><span class="ex">unique</span><span class="er">(</span><span class="ex">dat.samp</span><span class="va">$Geo_Location</span><span class="kw">))</span></span>
<span id="cb75-23"><a href="module-5-phylodynamics.html#cb75-23" tabindex="-1"></a></span>
<span id="cb75-24"><a href="module-5-phylodynamics.html#cb75-24" tabindex="-1"></a><span class="co">#setup colors for countries</span></span>
<span id="cb75-25"><a href="module-5-phylodynamics.html#cb75-25" tabindex="-1"></a><span class="ex">country.cols</span><span class="op">&lt;</span>-colorRampPalette<span class="er">(</span><span class="ex">brewer.pal</span><span class="er">(</span><span class="va">n</span><span class="op">=</span>8,name=<span class="st">&quot;Dark2&quot;</span><span class="kw">))(</span><span class="ex">n.c</span><span class="kw">)</span></span>
<span id="cb75-26"><a href="module-5-phylodynamics.html#cb75-26" tabindex="-1"></a><span class="ex">names</span><span class="er">(</span><span class="ex">country.cols</span><span class="kw">)</span><span class="op">&lt;</span>-order.c</span>
<span id="cb75-27"><a href="module-5-phylodynamics.html#cb75-27" tabindex="-1"></a></span>
<span id="cb75-28"><a href="module-5-phylodynamics.html#cb75-28" tabindex="-1"></a><span class="co">#factor and levels specifies order in legend</span></span>
<span id="cb75-29"><a href="module-5-phylodynamics.html#cb75-29" tabindex="-1"></a><span class="ex">dat.samp</span><span class="va">$Country</span><span class="op">&lt;</span>-factor<span class="er">(</span><span class="ex">dat.samp</span><span class="va">$Country</span><span class="ex">,</span> levels=order.c<span class="kw">)</span></span>
<span id="cb75-30"><a href="module-5-phylodynamics.html#cb75-30" tabindex="-1"></a></span>
<span id="cb75-31"><a href="module-5-phylodynamics.html#cb75-31" tabindex="-1"></a><span class="co">#plot country contribs</span></span>
<span id="cb75-32"><a href="module-5-phylodynamics.html#cb75-32" tabindex="-1"></a><span class="ex">ggplot</span><span class="er">(</span><span class="ex">dat.samp</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb75-33"><a href="module-5-phylodynamics.html#cb75-33" tabindex="-1"></a>  <span class="ex">geom_bar</span><span class="er">(</span><span class="ex">aes</span><span class="er">(</span><span class="va">x</span><span class="op">=</span>Country, <span class="va">fill</span><span class="op">=</span>Country<span class="kw">))</span><span class="ex">+</span></span>
<span id="cb75-34"><a href="module-5-phylodynamics.html#cb75-34" tabindex="-1"></a>  <span class="fu">theme_bw()</span><span class="ex">+</span></span>
<span id="cb75-35"><a href="module-5-phylodynamics.html#cb75-35" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="va">x</span><span class="op">=</span><span class="st">&quot;Country&quot;</span>, <span class="va">y</span><span class="op">=</span><span class="st">&quot;N sample&quot;</span>, <span class="va">fill</span><span class="op">=</span><span class="st">&quot;Country&quot;</span><span class="kw">)</span><span class="ex">+</span></span>
<span id="cb75-36"><a href="module-5-phylodynamics.html#cb75-36" tabindex="-1"></a>  <span class="ex">scale_fill_manual</span><span class="er">(</span><span class="va">values</span><span class="op">=</span>country.cols<span class="kw">)</span><span class="ex">+</span></span>
<span id="cb75-37"><a href="module-5-phylodynamics.html#cb75-37" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.text.x=element_text</span><span class="er">(</span><span class="va">angle</span><span class="op">=</span>45,vjust=1,hjust=1<span class="kw">)</span><span class="ex">,</span></span>
<span id="cb75-38"><a href="module-5-phylodynamics.html#cb75-38" tabindex="-1"></a>        <span class="ex">legend.position=</span><span class="st">&quot;none&quot;</span><span class="kw">)</span></span>
<span id="cb75-39"><a href="module-5-phylodynamics.html#cb75-39" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="ex">paste0</span><span class="er">(</span><span class="st">&quot;data/complete/samples_country.png&quot;</span><span class="kw">))</span></span>
<span id="cb75-40"><a href="module-5-phylodynamics.html#cb75-40" tabindex="-1"></a></span>
<span id="cb75-41"><a href="module-5-phylodynamics.html#cb75-41" tabindex="-1"></a><span class="co">#plot it over time</span></span>
<span id="cb75-42"><a href="module-5-phylodynamics.html#cb75-42" tabindex="-1"></a><span class="ex">ggplot</span><span class="er">(</span><span class="ex">dat.samp</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb75-43"><a href="module-5-phylodynamics.html#cb75-43" tabindex="-1"></a>  <span class="ex">geom_bar</span><span class="er">(</span><span class="ex">aes</span><span class="er">(</span><span class="va">x</span><span class="op">=</span>date, <span class="va">fill</span><span class="op">=</span>Country<span class="kw">))</span><span class="ex">+</span></span>
<span id="cb75-44"><a href="module-5-phylodynamics.html#cb75-44" tabindex="-1"></a>  <span class="fu">theme_bw()</span><span class="ex">+</span></span>
<span id="cb75-45"><a href="module-5-phylodynamics.html#cb75-45" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="va">x</span><span class="op">=</span><span class="st">&quot;Sample collection date&quot;</span>, <span class="va">y</span><span class="op">=</span><span class="st">&quot;N sample&quot;</span>, <span class="va">fill</span><span class="op">=</span><span class="st">&quot;Country&quot;</span><span class="kw">)</span><span class="ex">+</span></span>
<span id="cb75-46"><a href="module-5-phylodynamics.html#cb75-46" tabindex="-1"></a>  <span class="ex">scale_fill_manual</span><span class="er">(</span><span class="va">values</span><span class="op">=</span>country.cols<span class="kw">)</span></span>
<span id="cb75-47"><a href="module-5-phylodynamics.html#cb75-47" tabindex="-1"></a>  </span>
<span id="cb75-48"><a href="module-5-phylodynamics.html#cb75-48" tabindex="-1"></a><span class="co">#save it</span></span>
<span id="cb75-49"><a href="module-5-phylodynamics.html#cb75-49" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="ex">paste0</span><span class="er">(</span><span class="st">&quot;data/complete/samples_time_country.png&quot;</span><span class="kw">))</span></span></code></pre></div>
<p>If you are in command line R, enter q() to get out.</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>How many sequences were included from the UK?</li>
<li>What is the range of sample collection dates in the final dataset?</li>
<li>How many countries were represented in the final dataset?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>89</li>
<li>2020-09-20 to 2021-01-01</li>
<li>17 countries</li>
</ul></div></details>
<p><a name="phylodyn-mod5"></a></p>
</div>
</div>
<div id="maximum-likelihood-phylodynamic-analysis" class="section level3 hasAnchor">
<h3>4. Maximum likelihood phylodynamic analysis<a href="module-5-phylodynamics.html#maximum-likelihood-phylodynamic-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The overall goal of this lab is to estimate where and when B.1.1.7 most likely originated. To do this, we will mostly make use of the <a href="https://docs.nextstrain.org/projects/augur/en/stable/index.html">Augur</a> maximum likelihood tool suite, which powers the <a href="https://nextstrain.org/">NextStrain</a> website. Within augur, the key tool actually being used for most of our inferences is <a href="https://treetime.readthedocs.io/en/latest/">TreeTime</a> a likelihood based tool for inferring molecular clocks and ancestral traits.</p>
<p><strong>An overview of the basic usage of Augur (figure from the Augur documentation):</strong>
<img src="img/PGE_mod5/augur_analysis_sketch.webp" alt="augur workflow" /></p>
<p>The steps/functions we use are in this section of the lab are the following:</p>
<ul>
<li><strong>align</strong>: This step constructs a multiple sequence alignment from this subset of genomes.</li>
<li><strong>index</strong>: This step indexes the sequences, counting the frequency of nucleotides and gaps.</li>
<li><strong>filter</strong>: This step filters out sequences that have too many gaps or ambiguous nucleotides.</li>
<li><strong>tree</strong>: This step builds a phylogenetic tree, where branch lengths are measured in substitutions/site (a divergence tree).</li>
<li><strong>refine</strong>: This step constructs a time tree using our existing tree alongside collection dates of SARS-CoV-2 genomic samples (branch lengths are measured in time).</li>
<li><strong>traits</strong>: This step uses treetime’s mugration model to infer ancestral traits based on the tree. We will use this to infer ancestral host and geography.</li>
<li><strong>export</strong>: This step exports the data to be used by <a href="https://auspice.us/">Auspice</a>, a version of the visualization system used by NextStrain that lets you examine your own phylogenetic tree and metadata.</li>
</ul>
<hr />
<div id="construct-a-multiple-sequence-alignment-of-the-genomes-augur-align" class="section level4 hasAnchor">
<h4>4.1. Construct a multiple sequence alignment of the genomes (<code>augur align</code>)<a href="module-5-phylodynamics.html#construct-a-multiple-sequence-alignment-of-the-genomes-augur-align" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The first step is to construct a multiple sequence alignment of the genomes, which is required before building a phylogenetic tree. We will be using the command <code>augur align</code> to accomplish this task, but underneath this <a href="https://mafft.cbrc.jp/alignment/software/">mafft</a> runs to construct the alignment.</p>
<p>To build a new folder and construct the alignment, please run the following:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb76"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb76-1"><a href="module-5-phylodynamics.html#cb76-1" tabindex="-1"></a><span class="fu">mkdir</span> data/aligned</span>
<span id="cb76-2"><a href="module-5-phylodynamics.html#cb76-2" tabindex="-1"></a></span>
<span id="cb76-3"><a href="module-5-phylodynamics.html#cb76-3" tabindex="-1"></a><span class="ex">augur</span> align <span class="at">--nthreads</span> 2 <span class="at">--sequences</span> data/complete/b117_seq_complete_sampUK.fasta <span class="at">--reference-sequence</span> data/download/MN908947.3.fasta <span class="at">--output</span> data/aligned/b117_seq_complete_sampUK_aligned.fasta</span></code></pre></div>
<p>You should expect to see the following:</p>
<p><strong>Output</strong></p>
<pre><code>using mafft to align via:
    mafft --reorder --anysymbol --nomemsave --adjustdirection --thread 2 data/b117_seq_complete_sampUK_aligned.fasta.to_align.fasta 1&gt; data/b117_seq_complete_sampUK_aligned.fasta 2&gt; data/b117_seq_complete_sampUK_aligned.fasta.log 

    Katoh et al, Nucleic Acid Research, vol 30, issue 14
    https://doi.org/10.1093%2Fnar%2Fgkf436

WARNING: 1bp insertion at ref position 200 was due to &#39;N&#39;s or &#39;?&#39;s in provided sequences
WARNING: 1bp insertion at ref position 6337 was due to &#39;N&#39;s or &#39;?&#39;s in provided sequences
Trimmed gaps in MN908947.3|China|2019-12-26 from the alignment</code></pre>
<p>The meaning of each parameter:</p>
<ul>
<li>`–nthreads: Number of threads for the alignment.</li>
<li>`–sequences: The input set of sequences in FASTA format.</li>
<li>`–output: The output alignment, in FASTA format.</li>
<li>`–reference-sequence: The reference genome (MN908947.3==Wuhan-Hu-1).</li>
</ul>
<p>The reference sequence will be included in our alignment and <code>augur align</code> removes any insertions with respect to this reference genome (useful when identifying and naming specific mutations later on in the augur pipeline).</p>
<p>The aligned fasta file is a very similar format as the input file, but the difference is that sequences have been aligned (possibly by inserting gaps <code>-</code>). This also means that all sequences in <code>alignment.fasta</code> should have the same length (whereas sequences in <code>sequences.fasta</code>, which is not aligned, may have different lengths). When time allows, it is good practice to visually inspect the alignment in a viewer such as <a href="https://github.com/AliView/AliView">AliView</a>.</p>
</div>
<div id="index-sequences-augur-index" class="section level4 hasAnchor">
<h4>4.2. Index sequences (<code>augur index</code>)<a href="module-5-phylodynamics.html#index-sequences-augur-index" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This step is optional but useful to inspect sequence quality and quicken the subsequent filtering step, by counting the frequency of nucleotides and gaps.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb78"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb78-1"><a href="module-5-phylodynamics.html#cb78-1" tabindex="-1"></a><span class="ex">augur</span> index <span class="at">--sequences</span> data/aligned/b117_seq_complete_sampUK_aligned.fasta  <span class="at">--output</span>  data/aligned/b117_seq_complete_sampUK_aligned_sequence_index.tsv</span></code></pre></div>
<p>Inspect the index output in the ‘aligned’ subfolder.</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>Knowing that the genome size is 29,903 bp, do any samples appear to have more gaps or Ns than you would expect?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>At first glance, none appear to have &gt;~3000 gaps or Ns.</li>
</ul></div></details>
</div>
<div id="filter-out-low-quality-sequences-augur-filter" class="section level4 hasAnchor">
<h4>4.3. Filter out low quality sequences (<code>augur filter</code>)<a href="module-5-phylodynamics.html#filter-out-low-quality-sequences-augur-filter" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Filter seqs to a minimum length (–min-length) of 26913, which represents 90% of 29,903 sites with ACTG (i.e., max 10% ambig and gaps)</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb79"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb79-1"><a href="module-5-phylodynamics.html#cb79-1" tabindex="-1"></a><span class="fu">mkdir</span> data/filtered</span>
<span id="cb79-2"><a href="module-5-phylodynamics.html#cb79-2" tabindex="-1"></a></span>
<span id="cb79-3"><a href="module-5-phylodynamics.html#cb79-3" tabindex="-1"></a><span class="ex">augur</span> filter <span class="at">--nthreads</span> 2 <span class="at">--sequences</span> data/aligned/b117_seq_complete_sampUK_aligned.fasta <span class="at">--metadata</span> data/complete/b117_seq_complete_sampUK.tsv <span class="at">--sequence-index</span> data/aligned/b117_seq_complete_sampUK_aligned_sequence_index.tsv <span class="at">--min-length</span> 26913 <span class="at">--output-sequences</span> data/filtered/b117_seq_complete_sampUK_aligned_filt.fasta <span class="at">--output-metadata</span> data/filtered/b117_seq_complete_sampUK_filt.tsv </span></code></pre></div>
<p><strong>Output</strong></p>
<pre><code>1 strain was dropped during filtering
    1 had no metadata (this is the refence sequence)
178 strains passed all filters
</code></pre>
<p>Nice, all the sequences passed our quality filters (this is rarer than you might think).</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>Why is it important to remove low quality sequences prior to phylogenetic inference?</li>
<li>The choice of maximum 10% ambiguous and gaps seems (<em>is</em>) somewhat arbitrary. How would increasing or decreasing this threshold change the analysis?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>Sequences with excessive gaps or ambiguous sites may not have as
high of accuracy, and have less signal to contribute to the tree, which
could bias its placement.</li>
<li>Increasing the threshold would allow more sequences of lower quality
to pass through the filter. Decreasing the threshold is more
conservative.</li>
</ul></div></details>
</div>
<div id="mask-problematic-sites-in-the-alignment" class="section level4 hasAnchor">
<h4>4.4. Mask problematic sites in the alignment<a href="module-5-phylodynamics.html#mask-problematic-sites-in-the-alignment" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Not all sites in the genome are equally informative towards inferring their phylogenetic relationships. Some sites can be misleading and should be masked before building trees. The ends of alignments often exhibit low coverage and a higher rate of sequencing or mapping errors. In SARS-CoV-2, this includes positions 1–55 and 29804–29903. Homoplasic sites at which multiple recurrent mutations arise may reflect contamination, recurrent sequencing errors, or hypermutability. Even if they are accurate, they may not be representative of evolution across the remainder of the genome. Another example (not relevant here) would be sites at which drug resistance mutations occur. In HIV-1 phylodynamic analyses, codons of surveillance drug resistance mutations are masked. For further interest, see this <a href="https://virological.org/t/masking-strategies-for-sars-cov-2-alignments/480">justification for masking SARS-CoV-2 sequences</a>.</p>
<p>Note that augur also has a masking subcommand, but it requires the mask file to be in a different format, so we’re not using it here. Instead, we’re going to run a python script that masks sequences based on a vcf (Variant Call Format) file of problematic sites. In order to run this, we have to concatenate the reference sequence back into the fasta, which was filtered out by augur above. Then we’re going to run our masking script, and then remove that reference again.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb81"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb81-1"><a href="module-5-phylodynamics.html#cb81-1" tabindex="-1"></a><span class="fu">mkdir</span> data/masked</span>
<span id="cb81-2"><a href="module-5-phylodynamics.html#cb81-2" tabindex="-1"></a></span>
<span id="cb81-3"><a href="module-5-phylodynamics.html#cb81-3" tabindex="-1"></a><span class="co">## Concatenate the fasta with the reference fasta</span></span>
<span id="cb81-4"><a href="module-5-phylodynamics.html#cb81-4" tabindex="-1"></a><span class="fu">cat</span> data/filtered/b117_seq_complete_sampUK_aligned_filt.fasta data/download/MN908947.3.fasta <span class="op">&gt;</span> data/filtered/b117_seq_complete_sampUK_aligned_filt_ref.fasta</span>
<span id="cb81-5"><a href="module-5-phylodynamics.html#cb81-5" tabindex="-1"></a></span>
<span id="cb81-6"><a href="module-5-phylodynamics.html#cb81-6" tabindex="-1"></a><span class="co">## Run the masking python script</span></span>
<span id="cb81-7"><a href="module-5-phylodynamics.html#cb81-7" tabindex="-1"></a><span class="ex">python3</span> scripts/mask_alignment_using_vcf.py <span class="at">-i</span> data/filtered/b117_seq_complete_sampUK_aligned_filt_ref.fasta <span class="at">-o</span> data/masked/b117_seq_complete_sampUK_aligned_filt_ref_mask.fasta <span class="at">-v</span> scripts/problematic_sites_vcfv5.vcf <span class="at">-n</span> <span class="st">&quot;n&quot;</span> <span class="at">--reference_id</span> <span class="st">&quot;MN908947.3|China|2019-12-26&quot;</span></span>
<span id="cb81-8"><a href="module-5-phylodynamics.html#cb81-8" tabindex="-1"></a></span>
<span id="cb81-9"><a href="module-5-phylodynamics.html#cb81-9" tabindex="-1"></a><span class="co">## Remove the reference sequence from the alignment using augur filter</span></span>
<span id="cb81-10"><a href="module-5-phylodynamics.html#cb81-10" tabindex="-1"></a><span class="ex">augur</span> filter <span class="at">--sequences</span> data/masked/b117_seq_complete_sampUK_aligned_filt_ref_mask.fasta <span class="at">--metadata</span> data/filtered/b117_seq_complete_sampUK_filt.tsv <span class="at">--output-sequences</span> data/masked/b117_seq_complete_sampUK_aligned_filt_mask.fasta <span class="at">--output-metadata</span> data/masked/b117_seq_complete_sampUK_aligned_filt_mask.tsv</span></code></pre></div>
<p>The $cat command is a simple and useful tool to concatenate multiple fasta files together.</p>
<p>Let’s have a closer look at the parameters we used in the python script:</p>
<ul>
<li><code>-i</code>: Fasta input alignment to be masked (with the ref sequence in it)</li>
<li><code>-o</code>: Fasta output alignment with problematic sites masked</li>
<li><code>-v</code>: The vcf file of problematic sites from <a href="https://virological.org/t/masking-strategies-for-sars-cov-2-alignments/480">de Maio</a></li>
<li><code>-n</code>: the character to replace nucleotides at the masked sites</li>
<li><code>--reference-id</code>: the ID of the reference sequence in the input</li>
</ul>
</div>
<div id="optional" class="section level4 hasAnchor">
<h4>Optional<a href="module-5-phylodynamics.html#optional" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>Open and inspect the fasta files before and after masking.</li>
</ul>
</div>
<div id="build-a-maximum-liklihood-phylogenetic-tree-augur-tree" class="section level4 hasAnchor">
<h4>4.5. Build a maximum liklihood phylogenetic tree (<code>augur tree</code>)<a href="module-5-phylodynamics.html#build-a-maximum-liklihood-phylogenetic-tree-augur-tree" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The next step is to take the set of aligned genomes and build a phylogenetic tree (a divergence tree in units of substitutions per site). We will use <code>augur tree</code> for this, but underneath it (by default) runs <a href="http://www.iqtree.org/">iqtree</a>, which uses maximum likelihood to infer a phylogenetic tree.</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb82"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb82-1"><a href="module-5-phylodynamics.html#cb82-1" tabindex="-1"></a><span class="fu">mkdir</span> trees</span>
<span id="cb82-2"><a href="module-5-phylodynamics.html#cb82-2" tabindex="-1"></a></span>
<span id="cb82-3"><a href="module-5-phylodynamics.html#cb82-3" tabindex="-1"></a><span class="ex">augur</span> tree <span class="at">--nthreads</span> 4 <span class="at">--method</span> iqtree <span class="at">--substitution-model</span> GTR <span class="at">--alignment</span> data/masked/b117_seq_complete_sampUK_aligned_filt_mask.fasta <span class="at">--output</span> trees/b117_seq_complete_sampUK_aligned_filt_mask.tre</span></code></pre></div>
<p>Let’s check out the parameters</p>
<ul>
<li><code>--nthreads</code>: Number of threads</li>
<li><code>--method</code>: The tree inference method. We use the default, iqtree(3)</li>
<li><code>--substitution-model</code>: the underlying substitution model, here specified as generalized time reversible (GTR), which means that each nucleotide substitution rate in the matrix (ACTG vs ACTG) is separately estimated.</li>
<li><code>--alignment</code>: the input alignment (with masked sites)</li>
<li><code>--output</code>: the output tree (can be .tre or .nwk files)</li>
</ul>
<p>You should expect to see this (or a very similar) output:</p>
<p><strong>Output</strong></p>
<pre><code>Building a tree via:
    iqtree3 --threads-max 4 -s data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.fasta -m GTR --ninit 2 -n 2 --epsilon 0.05 -T AUTO --redo &gt; data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.iqtree.log
    Please use the corresponding citations according to
    https://iqtree.github.io/doc/Home#how-to-cite-iq-tree

Building original tree took 12.720907926559448 seconds
</code></pre>
<p>This produces as output a <code>.tre</code> file, which is a Newick format phylogenetic tree, which has been midpoint-rooted. You can load this file in a variety of phylogenetic tree viewers (such as <a href="http://phylo.io/" class="uri">http://phylo.io/</a>) but we will further refine this file to work with Auspice.</p>
<p>Another output file is <code>alignment-delim.iqtree.log</code>, which contains additional information from <a href="http://www.iqtree.org/">iqtree</a>. You can take a look at this file to get an idea of what <a href="http://www.iqtree.org/">iqtree</a> was doing by using <code>tail</code> (prints the last few lines of a file).</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb84"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb84-1"><a href="module-5-phylodynamics.html#cb84-1" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> 20 data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.iqtree.log</span></code></pre></div>
<p><strong>Output</strong></p>
<pre><code>Optimal log-likelihood: -44067.898
Rate parameters:  A-C: 0.24165  A-G: 1.07984  A-T: 0.21863  C-G: 0.31202  C-T: 3.23196  G-T: 1.00000
Base frequencies:  A: 0.299  C: 0.184  G: 0.196  T: 0.321
Parameters optimization took 1 rounds (0.019 sec)
BEST SCORE FOUND : -44067.898
Total tree length: 0.012

Total number of iterations: 2
CPU time used for tree search: 1.277 sec (0h:0m:1s)
Wall-clock time used for tree search: 0.408 sec (0h:0m:0s)
Total CPU time used: 29.925 sec (0h:0m:29s)
Total wall-clock time used: 12.023 sec (0h:0m:12s)

Analysis results written to: 
  IQ-TREE report:                data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.fasta.iqtree
  Maximum-likelihood tree:       data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.fasta.treefile
  Likelihood distances:          data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.fasta.mldist
  Screen log file:               data/masked/b117_seq_complete_sampUK_aligned_filt_mask-delim.fasta.log

Date and Time: Wed Nov 19 15:50:14 2025</code></pre>
<p>As iqtree uses a maximum likelihood approach, you will see that it will report the likelihood score of the optimal tree (reported as log-likelihoods since likelihood values are very small).</p>
<p><em>Note: For this lab we are not looking at branch support values for a tree, but for real-world analysis you may wish to look into including (ultra-fast) bootstrap support values or approximate likelihood ratio test values. This will give a measure of how well-supported each branch in the tree is by the alignment (often a number from 0 for little support to 100 for maximal support). Please see the <a href="http://www.iqtree.org/doc/Tutorial#assessing-branch-supports-with-ultrafast-bootstrap-approximation">IQTree documentation</a> for more details.</em></p>
</div>
<div id="inferring-timing-of-host-change-augur-refine" class="section level4 hasAnchor">
<h4>4.6. Inferring timing of host change (<code>augur refine</code>)<a href="module-5-phylodynamics.html#inferring-timing-of-host-change-augur-refine" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The tree output by <a href="http://www.iqtree.org/">iqtree</a> shows hypothetical evolutionary relationships between different SARS-CoV-2 genomes with branch lengths representing distances between different genomes (in units of <strong>substitutions/site</strong>, i.e., the predicted number of substitutions between genomes divided by the alignment length). However, other methods of measuring distance between genomes are possible. In particular we can incorporate the collection dates of the different SARS-CoV-2 genomes to infer a tree where branches are scaled according to the elapsed time and the dates of internal nodes are inferred. Such trees are called <strong>time trees</strong> or <strong>time-scaled phylogenies</strong>.</p>
<p>We will use <a href="https://treetime.readthedocs.io/en/latest/">TreeTime</a> invoked by <a href="https://docs.nextstrain.org/projects/augur/en/stable/index.html">augur</a> to infer a <strong>time tree</strong> from our phylogenetic tree using collection dates of the SARS-CoV-2 genomes stored in the metadata file. We will use the <code>augur refine</code> step to run TreeTime and perform some additional refinemint of the tree.</p>
<p>To do this, please run the following:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb86"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb86-1"><a href="module-5-phylodynamics.html#cb86-1" tabindex="-1"></a><span class="ex">augur</span> refine <span class="at">--alignment</span> data/masked/b117_seq_complete_sampUK_aligned_filt_mask.fasta <span class="at">--tree</span> trees/b117_seq_complete_sampUK_aligned_filt_mask.tre <span class="at">--metadata</span> data/masked/b117_seq_complete_sampUK_aligned_filt_mask.tsv <span class="at">--timetree</span> <span class="at">--divergence-units</span> mutations <span class="at">--output-tree</span> trees/b117_seq_complete_sampUK_aligned_filt_mask_timetree.tre <span class="at">--output-node-data</span> trees/refine.node.json <span class="at">--seed</span> 123 <span class="at">--stochastic-resolve</span> <span class="at">--root</span> least-squares <span class="at">--date-confidence</span> <span class="at">--clock-rate</span> 4.5e-4 <span class="at">--clock-std-dev</span> 0.2</span></code></pre></div>
<p>Once again, let’s have a closer look at the parameters:</p>
<ul>
<li><code>--alignment</code>: The alignment used to build the tree. Used to re-scale the divergence units.</li>
<li><code>--tree</code>: The starting input tree built using iqtree3.</li>
<li><code>--metadata</code>: The metadata which contains the SARS-CoV-2 genome names (in a column called <code>name</code>) and the sample collection dates (in a column named <code>date</code>).</li>
<li><code>--timetree</code>: Specifiy to build a time tree.</li>
<li><code>--divergence-units</code>: Convert the branch lengths of <strong>substitutions/site</strong> (<strong>mutations/site</strong>) to <strong>mutations</strong> (not needed to build a time tree, this is just used for visualizing the tree later on).</li>
<li><code>--output-tree</code>: The output Newick file containing the time tree.</li>
<li><code>--output-node-data</code>: Augur will store additional information here which will let us convert between time trees and substitution trees.</li>
<li><code>--seed</code>: Set a seed for reproducibility</li>
<li><code>--stochastic-resolve</code>: Resolve polytomies stochastically (forces a bifurcating tree)</li>
<li><code>--root</code>: choose root that optimizes the least squares for the root-to-tip regression
-date-confidence</li>
<li><code>--clock-rate</code>: We are specifying a relaxed molecular clock rate, with a mean of 4.5e-4 based on <a href="https://academic.oup.com/ve/article/8/2/veac080/6677185">Hill et al. 2022</a></li>
<li><code>--clock-std-dev</code>: This is the deviance on the relaxed clock, allowing for variable evolutionary rates across the tree. You could play with this parameter if you have time.</li>
</ul>
<p>You should expect to see the following as output:</p>
<p><strong>Output</strong></p>
<pre><code>augur refine is using TreeTime version 0.11.4

4.28    TreeTime.reroot: with method or node: least-squares

4.28    TreeTime.reroot: rerooting will account for covariance and shared ancestry.

5.32    ###TreeTime.run: INITIAL ROUND

9.24    TreeTime.reroot: with method or node: least-squares

9.24    TreeTime.reroot: rerooting will account for covariance and shared ancestry.

9.41    ###TreeTime.run: rerunning timetree after rerooting

13.59   ###TreeTime.run: ITERATION 1 out of 2 iterations

21.88   ###TreeTime.run: ITERATION 2 out of 2 iterations

55.09   ###TreeTime.run: FINAL ROUND - confidence estimation via marginal
        reconstruction
Inferred a time resolved phylogeny using TreeTime:
    Sagulenko et al. TreeTime: Maximum-likelihood phylodynamic analysis
    Virus Evolution, vol 4, https://academic.oup.com/ve/article/4/1/vex042/4794731

updated tree written to trees/b117_seq_complete_sampUK_aligned_filt_mask_timetree.tre
node attributes written to refine.node.json</code></pre>
<p>As output, the file <code>output-tree trees/b117_seq_complete_sampUK_aligned_filt_mask_timetree.tre</code> will contain the time tree while the file <code>trees/refine.node.json</code> contains additional information about the tree. If you want, download the time tree and view it in <a href="https://icytree.org/">icytree</a>.</p>
<div id="optional-1" class="section level5 hasAnchor">
<h5>Optional<a href="module-5-phylodynamics.html#optional-1" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>If you have spare time at the end of the lab, consider returning to conduct a sensitivity analysis on the specified clock rate and variance.</p>
<hr />
</div>
</div>
<div id="infer-ancestral-states-for-host-and-location-augur-traits" class="section level4 hasAnchor">
<h4>4.7. Infer ancestral states for host and location (<code>augur traits</code>)<a href="module-5-phylodynamics.html#infer-ancestral-states-for-host-and-location-augur-traits" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We will now be using treetime’s <code>mugration</code> model to reconstruct discrete ancestral states. In other words, we are trying to use the tree and genome metadata to reconstruct the most likely country at each of the internal nodes in the tree. It is called a mugration model because it is fundamentally estimating migration between “demes” (in this case, countries) as if it were a substitution rate matrix, commonly referred to as <em>mu</em>. See <a href="https://pubmed.ncbi.nlm.nih.gov/19779555/">Lemey et al. 2009</a> if you’re interested in learning more.</p>
<p>Side bar: discrete phylogeography, ancestral state reconstruction, ancestral trait inference, and ancestral character estimation can all be used interchangeably in this context.</p>
<p>As with the other analyses, augur makes this process very convenient:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb88"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb88-1"><a href="module-5-phylodynamics.html#cb88-1" tabindex="-1"></a><span class="ex">augur</span> traits <span class="at">--tree</span> trees/b117_seq_complete_sampUK_aligned_filt_mask_timetree.tre <span class="at">--metadata</span> data/masked/b117_seq_complete_sampUK_aligned_filt_mask.tsv <span class="at">--columns</span> Country Geo_Location <span class="at">--output-node-data</span> trees/trait.node.json</span></code></pre></div>
<p>The parameters we used are:</p>
<ul>
<li><code>--tree</code>: The time-calibrated tree we inferred in step 4</li>
<li><code>--metadata</code>: The metadata file with information about the genomes in the tree</li>
<li><code>--columns</code>: The columns in the metadata file we want to infer ancestral states for, in this case country and geo_location (more specific info).</li>
<li><code>--output-node-data</code>: Augur stores the additional information related to the ancestral trait inference for later visualisation.</li>
</ul>
<p>You should expect to see the following as output:</p>
<p><strong>Output</strong></p>
<pre><code>augur traits is using TreeTime version 0.11.4
Assigned discrete traits to 178 out of 178 taxa.

Assigned discrete traits to 178 out of 178 taxa.

Inferred ancestral states of discrete character using TreeTime:
    Sagulenko et al. TreeTime: Maximum-likelihood phylodynamic analysis
    Virus Evolution, vol 4, https://academic.oup.com/ve/article/4/1/vex042/4794731

results written to trait.node.json</code></pre>
<hr />
</div>
<div id="package-up-data-for-auspice-visualisation-augur-export" class="section level4 hasAnchor">
<h4>4.8. Package up data for Auspice visualisation (<code>augur export</code>)<a href="module-5-phylodynamics.html#package-up-data-for-auspice-visualisation-augur-export" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We will be using <a href="https://auspice.us/">Auspice</a> to visualize the tree alongside our metadata. To do this, we need to package up all of the data we have so far into a special file which can be used by Auspice. To do this, please run the following command:</p>
<p><strong>Commands</strong></p>
<div class="sourceCode" id="cb90"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb90-1"><a href="module-5-phylodynamics.html#cb90-1" tabindex="-1"></a><span class="ex">augur</span> export v2 <span class="at">--tree</span> trees/b117_seq_complete_sampUK_aligned_filt_mask_timetree.tre <span class="at">--node-data</span> trees/refine.node.json trees/trait.node.json <span class="at">--maintainers</span> <span class="st">&quot;CBW-IDE-2025&quot;</span> <span class="at">--title</span> <span class="st">&quot;Module 5 Practical&quot;</span> <span class="at">--output</span> trees/analysis-package.json <span class="at">--geo-resolutions</span> Country</span></code></pre></div>
<p>The parameters:</p>
<ul>
<li><code>--tree</code>: The time-calibrated tree</li>
<li><code>--node-data refine.node.json trait.node.json</code>: The 2 jsons created during the temporal and ancestral trait inference</li>
<li><code>--title "Module 5 Practical"</code>: A title for the auspice page, you can make this anything you want.</li>
<li><code>--maintainers "CBW-IDE-2025"</code>: A name for who is responsible for this analysis, useful for later, can make this anything you want.</li>
<li><code>--geo-resolutions Country</code>: Spatial resolution on which to plot the data. In this case we want country</li>
<li><code>--output analysis-package.json</code>: The key output file we want to generate. The file <code>trees/analysis-package.json</code> contains the tree with different branch length units (time and substitutions), our inferred ancestral traits, as well as additional data.</li>
</ul>
<p>You should expect to see the following as output:</p>
<p><strong>Output</strong></p>
<pre><code>Trait &#39;Geo_Location&#39; was guessed as being type &#39;categorical&#39;. Use a &#39;config&#39; file if you&#39;d like to set this yourself.
Trait &#39;Country&#39; was guessed as being type &#39;categorical&#39;. Use a &#39;config&#39; file if you&#39;d like to set this yourself.
Validating produced JSON
Validating schema of &#39;trees/analysis-package.json&#39;...
Validating that the JSON is internally consistent...
Validation of &#39;trees/analysis-package.json&#39; succeeded.</code></pre>
<hr />
<p><a name="visualize-tree-mod5"></a></p>
</div>
</div>
<div id="visualizing-the-phylodynamic-analysis" class="section level3 hasAnchor">
<h3>5. Visualizing the phylodynamic analysis<a href="module-5-phylodynamics.html#visualizing-the-phylodynamic-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we’ve constructed and packaged up a tree (<code>analysis-package.json</code>), we can visualize this data alongside the data augur has extracted from our metadata using augur and <a href="https://auspice.us/">Auspice</a>.</p>
<p>Note: I’ve had issues recently getting this to work in firefox so recommend using a chromium-based browser if you have issues.</p>
<hr />
<div id="load-data-into-auspice" class="section level4 hasAnchor">
<h4>5.1. Load data into Auspice<a href="module-5-phylodynamics.html#load-data-into-auspice" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To do this, please navigate to <a href="http://IP-ADDRESS/module5/" class="uri">http://IP-ADDRESS/module5/</a> and download the files <code>trees/analysis-package.json</code> and <code>data/masked/b117_seq_complete_sampUK_aligned_filt_mask.tsv</code> to your computer (if the link does not download you can <strong>Right-click</strong> and select <strong>Save link as…</strong>).</p>
<p>Next, navigate to <a href="https://auspice.us/" class="uri">https://auspice.us/</a> and drag the file <code>analysis-package.json</code> onto the page.</p>
<hr />
</div>
<div id="explore-data" class="section level4 hasAnchor">
<h4>5.2. Explore data<a href="module-5-phylodynamics.html#explore-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Spend some time to explore the data and get used to the Auspice interface. Try switching between different <strong>Tree Layouts</strong>, <strong>Branch Lengths</strong>, or colouring the tree by different criteria in the metadata table. This is always worth doing a bit before diving into analyses. Scroll down and hit download to save the outputs of this for future reference. The nexus trees contain the internal node states along with the tree, so is readable into R and FigTree if you want to customize the tree more or analyze ancestral state transitions to identify sublineages or calculate dispersion rates etc.</p>
<hr />
</div>
<div id="examine-the-molecular-clock" class="section level4 hasAnchor">
<h4>5.3. Examine the molecular clock<a href="module-5-phylodynamics.html#examine-the-molecular-clock" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now we are going to look more at the Branch length “TIME” and “DIVERGENCE” option.</p>
<p>First, let’s look at the divergence tree and use the x-axis to work out how many mutations there are on certain branches. We can use the x-axis in the “RECTANGULAR” layout of the “DIVERGENCE” tree to work this out or we can directly move our cursor over the node/branch of interest to have a pop-up with the details.</p>
<p><strong>The mutation tree:</strong>
<img src="img/PGE_mod5/auspice_muts.png" alt="auspice mutation tree" /></p>
<p>Now, we can do the same thing if we switch to “TIME” and look at the approximate inferred dates of the internal nodes of the tree. We can also mouse over to get the exact pop-up dates.</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>What was the maximum number of mutations from the root to any tip?</li>
<li>What was the most likely date of the root (AKA the <strong>time of the most recent common ancestor (TMRCA)</strong>) ? Confidence interval?</li>
<li>How do you interpret the root of this phylogeny? Is it representative of the entire B.1.1.7 lineage?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>18 (to tip MW598424.1/Ghana/2020-12-22). This tip could be
considered a temporal outlier and excluded from the analysis.</li>
<li>2020-08-21, but the confidence interval is wide and lower bound is
unrealistic (2016-05-31 - 2020-09-12)</li>
<li>This root is the most recent common ancestor (MRCA) of a random
sample of early B.1.1.7 sequences, downsampling the UK samples to an
equal number as non-UK samples. It may not be fully representative of
the full diversity of early B.1.1.7, nor later evolution of
B.1.1.7.</li>
</ul></div></details>
<p>Now, have a look at the “CLOCK” layout/figure found in the options bar on the left of your screen. This brings up a <strong>root-to-tip regression</strong> from the divergence tree (i.e., how many mutations from root to tip, versus collection date). You can click on individual points (including the upper outlier from Ghana) to see the average substitution rate for that tip. for that tip.</p>
<p><strong>The root-to-tip regression plot:</strong>
<img src="img/PGE_mod5/auspice_rtt.png" alt="auspice root-to-tip" /></p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>What is the slope of the root-to-tip regression (in subs/year; subs/site/year)? What does it represent?</li>
<li>Does the trend line for the root-to-tip regression look like a good fit with good temporal signal?</li>
<li>Given the data and trendline, how accurate do you think the inferred root date is?</li>
<li>How might you estimate the uncertainty in these estimates?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>14.9 substitutions per year. Or (ignoring masked sites),
14.9/29903=0.0004982778=5.0 subs/site/year. It represents the strict
molecular clock rate.</li>
<li>Aside from the outlier, it looks like a good fit (an R-squared value
should really be reported) with reasonable temporal signal.</li>
<li>The confidence interval on the root is wide, we have a relatively
small (singular) sample, but the linear fit is decent and we included
early samples with higher probabilities, so I’d say medium to low
accuracy for the mean estimate.</li>
<li>Multiple subsamples of higher sample sizes. Bootstrapping/ultrafast
bootstrapping. Multiple clock rate assumptions when inferring the time
tree.</li>
</ul></div></details>
<hr />
</div>
<div id="examine-the-ancestral-state-reconstruction" class="section level4 hasAnchor">
<h4>5.4. Examine the ancestral state reconstruction<a href="module-5-phylodynamics.html#examine-the-ancestral-state-reconstruction" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can colour the tree and rename the tree tip labels using the metadata to reflect the ancestral state reconstruction. The internal nodes/branches will be coloured based on the inferred ancestral state, in this case Country or Geo_location. Investigate the ancestral states at internal nodes throughout the tree including at the root. Use the play button to visualize the dispersion globally.</p>
<p><strong>Here is a look at what you should be seeing in auspice when you change branches to be colored by Country and choose the multi-panel layout.</strong>
<img src="img/PGE_mod5/auspice_geo.png" alt="auspice phylogeography" /></p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>Based on this analysis, where was the root of B.1.1.7 likely to have occurred?</li>
<li>How do you think sequence subsampling might have impacted this analysis?</li>
<li>What additional data would give you more confidence in asserting where this lineage was likely to have originated?</li>
<li>How could you use the output of this analysis to quantify viral dispersion?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>The United Kingdom</li>
<li>We included far fewer sequences from the UK than were available, and
still inferred it as the likely origin. It would be worth comparing
other sampling schemes to validate this, but we have high confidence
that this was the likely origin, given the sequences publicly
available.</li>
<li>More early B.1.1.7 sequences from other countries would be
informative. Also, including close parental or sister lineages (like B.1
similar to B.1.1.7) of similar time/place could add resolution to the
analysis.</li>
<li>Quantify importation or exportation rates by country through time.
You could also incorporate population size, connectivity (mobility
data), geographical distance, and other factors to evaluate (or control
for) factors associated with elevated dispersal.</li>
</ul></div></details>
<hr />
<p>That’s that for our maximum likelihood phylogeographic inference. Now, we’re going to compare those results to a Bayesian approach.</p>
<p><a name="delphy-mod5"></a></p>
</div>
</div>
<div id="delphy-analysis" class="section level3 hasAnchor">
<h3>6. Delphy analysis<a href="module-5-phylodynamics.html#delphy-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://delphy.fathom.info/">Delphy</a> is a newly released Bayesian phylogenetics tool designed to be fast, scalable, accessible, and accurate. Delphy is faster than other Bayesian tools because it uses Explicit Mutation-Annotated Trees (EMAT), which are hypotheses of how an ancestral sequence replicated and evolved, mutation by mutation, into the observed dated sequences. Check out their <a href="https://www.biorxiv.org/content/10.1101/2025.03.25.645253v1.full.pdf">pre-print</a> if you want more details.</p>
<p>In Bayesian phylogenetic (and phylodynamic) inference, many models (made up of trees and parameters) are sampled to generate posterior probability distributions, and our estimates will reflect the <strong>highest posterior density (HPD)</strong>. Bayesian analyses differ from max likelihood approaches in that they incorporate uncertainty from the tree inference and sampling process, and rely on specification of prior probabilities for the parameters to be explored using <strong>Markov Chain Monte Carlo (MCMC)</strong>, which is basically a hill-climbing algorithm that suggests and compares models compared to the previous step. Model convergence is assessed using minimum <a href="https://github.com/fathominfo/delphy-web/wiki/Effective-Sample-Size"><strong>effective sampling size (ESS)</strong></a> reflects and in robust analyses is usually ~200 after removing the burn-in. Burn-in are steps that you don’t include in the final posterior distributions because they represent the model stabilization period. In a robust Bayesian analyses, multiple <strong>MCMC chains</strong> are run, and then combined after burn-in.</p>
<p>The tree displayed reflects a <strong>maximum clade credibility</strong> tree, which is a single tree with the highest posterior probability across the Bayesian posterior distribution of trees. Branch width represents the credibility threshold, more commonly referred to as <strong>posterior support or probability</strong>, which indicate the strength of evidence for clades, given the data and model used.</p>
<div id="modify-fasta-names-then-load-the-renamed-fasta-into-delphy" class="section level4 hasAnchor">
<h4>6.1. Modify fasta names, then load the renamed fasta into Delphy<a href="module-5-phylodynamics.html#modify-fasta-names-then-load-the-renamed-fasta-into-delphy" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Delphy wants sequence names to include info separated by “|” not “/”, so we’ll start by running a simple R script to rename the relevant fasta and metadata.</p>
<p>Run this Rscript in the command line. You can go to scripts/convert_names.R to have a look at what this script does, but it’s pretty simple.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb92-1"><a href="module-5-phylodynamics.html#cb92-1" tabindex="-1"></a><span class="ex">Rscript</span> scripts/convert_names.R <span class="st">&quot;data/masked/b117_seq_complete_sampUK_aligned_filt_mask.tsv&quot;</span> <span class="st">&quot;data/masked/b117_seq_complete_sampUK_aligned_filt_mask.fasta&quot;</span></span></code></pre></div>
<p>The renamed files output goes to a newly created folder, <code>data/delphy</code>.</p>
<p>Now, open the <a href="delphy.fathom.info">Delphy web browser</a> and drag/drop or upload the fasta file <code>data/delphyb117_seq_complete_sampUK_aligned_filt_mask_delphy.fasta</code>.</p>
</div>
<div id="modify-priors-by-clicking-on-advanced-options" class="section level4 hasAnchor">
<h4>6.2. Modify priors by clicking on advanced options<a href="module-5-phylodynamics.html#modify-priors-by-clicking-on-advanced-options" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Leave the sampling frequency to sample 1 tree per 1,000,000 steps. You could change this to sample more frequently if you want, but this might not improve the speed of convergence if trees are too similar to each other.</p>
<p>Let’s modify the priors. It is important to assess and modify priors in Bayesian analyses to reflect your <em>prior</em> expectations and narrow the parameter space explored. Improper priors not only slow down convergence, but can bias analyses.</p>
<ul>
<li>For mutation rate:
<ul>
<li>enable site rate heterogeneity</li>
<li>fix overall (mean) mutation rate</li>
<li>set initial mutation rate to 45e-5 s/s/y 4.5e-4 (Hill et al. 2022)</li>
</ul></li>
<li>Population model
<ul>
<li>Choose skygrid (these are both coalescent tree priors. Delphy does not have birth-death models like BEAST does.)</li>
<li>model start date 2020-01-01 (the earliest this lineage root could have occurred)</li>
<li>split into 10 intervals</li>
<li>keep other parameters the same (could play with advanced skygrid parameters)
<ul>
<li><a href="https://github.com/broadinstitute/delphy/wiki/Flexible-population-models">Further reading on the choice of population model</a></li>
</ul></li>
</ul></li>
</ul>
<p><strong>Here is a look at the advanced options you should be modifying. This should also have had split into 10 intervals.</strong>
<img src="img/PGE_mod5/delphy_prior.png" alt="delphy priors" /></p>
<p>Before we start the run, go to Customize in the toolbar, then add the metadata file <code>data/delphyb117_seq_complete_sampUK_aligned_filt_mask_delphy.tsv</code> by drag/drop then in Color system, click metadata, color by Country, then select all.</p>
<p><strong>Delphy customize options:</strong>
<img src="img/PGE_mod5/delphy_cust.png" alt="delphy add metadata" /></p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li>what biases may we have imposed through our selection of priors?</li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li>We may bias our estimates of the root date by specifying an improper
mean molecular clock rate. The skygrid may not fit well with 10
categories (not enough data in early period). The underlying population
may be better described by the exponential population distribution. Any
misspecified or over-generalized prior has the potential to bias the
analysis or underestimate the true uncertainty in the estimates.</li>
</ul></div></details>
</div>
<div id="run-the-analysis-and-interpret-output" class="section level4 hasAnchor">
<h4>6.3. Run the analysis and interpret output<a href="module-5-phylodynamics.html#run-the-analysis-and-interpret-output" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Hit play and watch in <em>statistical awe</em> as trees sampled from the posterior are displayed in real-time, alongside model parameters including the number of mutations, ln(posterior), total evolutionary time (tree height), and effective population size in years. We can start to investigate the output before the run is complete (or you can upload the precomputed-analysis/trees/delphy_out/.delphy file directly to the Delphy home page to analyze a more complete analysis of &gt;300 million steps).</p>
<p><strong>Here is a look at the more complete run, including a sampled tree, the MCC tree, and the posterior distribution of some key model parameters.</strong>
<img src="img/PGE_mod5/delphy_mcc.png" alt="delphy mcc" /></p>
<p>Although Delphy does not seem very good at estimating the ESS and assessing convergence, a good sign of convergence is when the posterior distributions of parameters appear as “fuzzy caterpillars” (converging around similar values, as opposed to jumping around).</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li><p>How do you interpret the change in effective population size through time?</p></li>
<li><p>Did your model converge? How can you tell?</p></li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li><p>The effective population size is low and then increases rapidly
(exponentially?) as this lineage transmitted broadly in the UK and
globally, but then appears to decrease at the end of the study perio,
but this likely reflects lower sampling rate towards the end of the
period.</p></li>
<li><p>After removing burn-in, the ESS for the parameters shown are
below 200, so probably not. We would have to run this for more steps,
and maybe for multiple chains that we could combine using a tool like <a
href="https://www.beast2.org/tracer-2/">Tracer</a>.</p></li>
</ul></div></details>
<p>NEXT, let’s click on Lineages in the upper tool bar. In the bottom left corner, you can change branch coloring between Credibility and Metadata (Country). You can also change the credibility threshold (posterior support) for nodes’/branches’ shading in the tree displayed.</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li><p>Is the first bifurcation at the root well-supported?</p></li>
<li><p>What is the median posterior root date?</p></li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li><p>No, the credibility threshold is &lt;50%.</p></li>
<li><p>Median posterior root date ~19 July 2020. Normally a 95% HPD
should be reported, but Delphy does not show that directly. If you have
time/interest, you can read the exported data into R and estimate the
uncertainty.</p></li>
</ul></div></details>
<p>And now click on Mutations in the upper tool bar. Click on recurrent mutations, then A17615G. This mutation may have evolved independently multiple times, though it was only found in 70 genomes. It is a non-synonymous mutation in the ORF1b gene, resulting in an amino acid change of Lysine to Arginine (K1383R) in the nsp10 protein. Explore the other mutations that occurred in the phylogeny.</p>
<div class="callout-orange">
<span class="callout-icon"><i class="fas fa-fire" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Questions</p>
</div>
<ul>
<li><p>Where does the Delphy analysis suggest the root of this B.1.1.7 phylogeny occurred?</p></li>
<li><p>Does the Bayesian analysis corroborate the maximum likelihood analysis in terms of the time and place of origin of B.1.1.7?</p></li>
</ul>
</div>
</div>
<details class="callout-dropdown callout-green"><summary class="callout-dropdown-summary"><i class="fas fa-circle-check" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">Answers</span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><ul>
<li><p>The UK</p></li>
<li><p>Yes! Both the Bayesian and ML analyses suggest this lineage
originated around August 20, 2020 in the UK. This is very close to the
estimate from <a
href="https://academic.oup.com/ve/article/8/2/veac080/6677185">Hill et
al. 2022</a> of a “TMRCA of the Alpha clade estimated at 28 August 2020
(95 % HPD: 15th August 2020 to 9th September 2020)”.</p></li>
</ul></div></details>
</div>
<div id="export-results" class="section level4 hasAnchor">
<h4>6.4. Export results<a href="module-5-phylodynamics.html#export-results" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To wrap this Delphy analysis up, go to Customize, then in the bottom left, export the Tree as png and Newick file, and the Delphy run. The Delphy run file can be dropped back into Delphy or read into R. You can also generate input/output to BEAST to run a more <em>bespoke</em> Bayesian analysis.</p>
<p><a name="end-mod5"></a></p>
</div>
</div>
<div id="end-of-lab" class="section level3 hasAnchor">
<h3>7. End of lab<a href="module-5-phylodynamics.html#end-of-lab" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You’ve made it to the end of the lab. Awesome job.</p>
<details class="callout-dropdown callout-orange"><summary class="callout-dropdown-summary"><i class="fas fa-fire" aria-hidden="true"></i>&nbsp;&nbsp;&nbsp;<span class="callout-dropdown-title">You deserve a <em>little treat</em></span> <span class="callout-dropdown-caret"><i class="fas fa-chevron-right" aria-hidden="true"></i></span></summary><div class="callout-dropdown-content"><figure>
<img src="img/PGE_mod5/end.png" alt="treat" />
<figcaption aria-hidden="true">treat</figcaption>
</figure></div></details>
<p>If you have some extra time, you can explore Auspice or Delphy further, or see how the results differ if you change the subsampling scheme or the clock rate. Another <em>fun</em> activity would be to read the Delphy output into R and summarize the median and 95% highest posterior distribution of the root date,</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-4-outbreak-analysis.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": false,
    "twitter": false,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca//blob/main/050-module-5.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
